--- src/Terraria/Terraria/GameContent/UI/BigProgressBar/BigProgressBarHelper.cs
+++ src/TerraAngel/Terraria/GameContent/UI/BigProgressBar/BigProgressBarHelper.cs
@@ -1,3 +_,4 @@
+using System;
 using Microsoft.Xna.Framework;
 using Microsoft.Xna.Framework.Graphics;
 using ReLogic.Graphics;
@@ -28,7 +_,7 @@
         int verticalFrames = 6;
         Rectangle value2 = value.Frame(1, verticalFrames, 0, 3);
         Color color = Color.White * 0.2f;
-        float num = lifeAmount / lifeMax;
+        float num = Math.Clamp(lifeAmount / lifeMax, 0f, 1f);
         int num2 = (int)((float)p.X * num);
         num2 -= num2 % 2;
         Rectangle value3 = value.Frame(1, verticalFrames, 0, 2);
