--- src/Terraria/ReLogic/OS/OSX/OsxPlatform.cs
+++ src/TerraAngel/ReLogic/OS/OSX/OsxPlatform.cs
@@ -1,19 +_,20 @@
 using System;
 using ReLogic.Localization.IME;
+using ReLogic.OS.FNA;
 
 namespace ReLogic.OS.OSX;
 internal class OsxPlatform : Platform
 {
     public OsxPlatform() : base(PlatformType.OSX)
     {
-        RegisterService((IClipboard)new Clipboard());
+        RegisterService((IClipboard)new FNAClipboard());
         RegisterService((IPathService)new PathService());
-        RegisterService((IWindowService)new WindowService());
+        RegisterService((IWindowService)new FNAWindow());
         RegisterService((IImeService)new UnsupportedPlatformIme());
     }
 
     public override void InitializeClientServices(IntPtr windowHandle)
     {
-        RegisterService((IImeService)new FnaIme());
+        RegisterService((IImeService)new FNAIme());
     }
 }
