--- src/Terraria/ReLogic/OS/OSX/OsxPlatform.cs
+++ src/TerraAngel/ReLogic/OS/OSX/OsxPlatform.cs
@@ -7,7 +_,7 @@
     {
         public OsxPlatform() : base(PlatformType.OSX)
         {
-            RegisterService((IClipboard)new Clipboard());
+            RegisterService((IClipboard)new FNAClipboard());
             RegisterService((IPathService)new PathService());
             RegisterService((IWindowService)new WindowService());
             RegisterService((IImeService)new UnsupportedPlatformIme());
@@ -15,7 +_,7 @@
 
         public override void InitializeClientServices(IntPtr windowHandle)
         {
-            RegisterService((IImeService)new FnaIme());
+            RegisterService((IImeService)new TerraAngel.Input.RealFNAIme());
         }
     }
 }
