--- src/Terraria/ReLogic/Threading/FastParallel.cs
+++ src/TerraAngel/ReLogic/Threading/FastParallel.cs
@@ -52,8 +_,6 @@
             num2 = 1;
         }
 
-        ThreadPriority priority = Thread.CurrentThread.Priority;
-        Thread.CurrentThread.Priority = ThreadPriority.Highest;
         int num3 = num / num2;
         int num4 = num % num2;
         CountdownEvent countdownEvent = new CountdownEvent(num2);
@@ -80,11 +_,7 @@
             }
         }
 
-        while (countdownEvent.CurrentCount != 0)
-        {
-        }
-
-        Thread.CurrentThread.Priority = priority;
+        countdownEvent.Wait();
     }
 
     private static void InvokeTask(object context)
