--- src/Terraria/Terraria/Collision.cs
+++ src/TerraAngel/Terraria/Collision.cs
@@ -1422,25 +_,6 @@
     public static bool GetWaterLine(int X, int Y, out float waterLineHeight)
     {
         waterLineHeight = 0f;
-        if (Main.tile[X, Y - 2] == null)
-        {
-            Main.tile[X, Y - 2] = new Tile();
-        }
-
-        if (Main.tile[X, Y - 1] == null)
-        {
-            Main.tile[X, Y - 1] = new Tile();
-        }
-
-        if (Main.tile[X, Y] == null)
-        {
-            Main.tile[X, Y] = new Tile();
-        }
-
-        if (Main.tile[X, Y + 1] == null)
-        {
-            Main.tile[X, Y + 1] = new Tile();
-        }
 
         if (Main.tile[X, Y - 2].liquid > 0)
         {
