--- src/Terraria/Terraria/Item.cs
+++ src/TerraAngel/Terraria/Item.cs
@@ -326,20 +_,7 @@
         return $"{{Name: \"{Name}\" NetID: {netID} Stack: {stack}";
     }
 
-    private bool CanHavePrefixes()
-    {
-        if (type != 0 && maxStack == 1)
-        {
-            if (damage <= 0)
-            {
-                return accessory;
-            }
-
-            return true;
-        }
-
-        return false;
-    }
+    private bool CanHavePrefixes() => true;
 
     public bool Prefix(int prefixWeWant)
     {
@@ -890,7 +_,9 @@
         return false;
     }
 
+    public string AffixName() => TerraAngel.Hooks.Hooks.DrawHooks.AffixNameHook((x) => x.AffixName(), this);
+
-    public string AffixName()
+    private string _AffixName()
     {
         if (prefix < 0 || prefix >= Lang.prefix.Length)
         {
@@ -49372,12 +_,15 @@
             velocity = Collision.TileCollision(position, velocity, width, height);
         }
 
+        if (Main.tile.InWorldAndLoaded(((int)(position.X / 16)), ((int)(position.Y / 16))))
+        {
         Vector4 vector2 = Collision.SlopeCollision(position, velocity, width, height, gravity);
         position.X = vector2.X;
         position.Y = vector2.Y;
         velocity.X = vector2.Z;
         velocity.Y = vector2.W;
         Collision.StepConveyorBelt(this, 1f);
+        }
     }
 
     private void GetPickedUpByMonsters_Special(int i)
