--- src/Terraria/Terraria/Item.cs
+++ src/TerraAngel/Terraria/Item.cs
@@ -320,20 +_,7 @@
         return $"{{Name: \"{Name}\" NetID: {netID} Stack: {stack}";
     }
 
-    private bool CanHavePrefixes()
-    {
-        if (type != 0 && maxStack == 1)
-        {
-            if (damage <= 0)
-            {
-                return accessory;
-            }
-
-            return true;
-        }
-
-        return false;
-    }
+    private bool CanHavePrefixes() => true;
 
     public bool Prefix(int prefixWeWant)
     {
@@ -49290,6 +_,7 @@
         velocity.X = vector2.Z;
         velocity.Y = vector2.W;
         Collision.StepConveyorBelt(this, 1f);
+        if (!Main.tile.IsTileInLoadedSection(((int)(position.X / 16)), ((int)(position.Y / 16)))) velocity = Vector2.Zero;
     }
 
     private void GetPickedUpByMonsters_Special(int i)
