--- src/Terraria/Terraria/Main.cs
+++ src/TerraAngel/Terraria/Main.cs
@@ -3,7 +_,6 @@
 using System.Collections.Concurrent;
 using System.Collections.Generic;
 using System.Diagnostics;
-using System.Drawing;
 using System.IO;
 using System.Linq;
 using System.Reflection;
@@ -12,7 +_,6 @@
 using System.Text.RegularExpressions;
 using System.Threading;
 using System.Threading.Tasks;
-using System.Windows.Forms;
 using Microsoft.Win32;
 using Microsoft.Xna.Framework;
 using Microsoft.Xna.Framework.Content;
@@ -23,8 +_,8 @@
 using ReLogic.Graphics;
 using ReLogic.Localization.IME;
 using ReLogic.OS;
-using ReLogic.Peripherals.RGB;
 using ReLogic.Utilities;
+using TerraAngel;
 using Terraria.Achievements;
 using Terraria.Audio;
 using Terraria.Chat;
@@ -203,8 +_,6 @@
     public static bool CaptureModeDisabled = false;
     public bool unityMouseOver;
     public static Main instance;
-    public static ChromaEngine Chroma;
-    public static ChromaHotkeyPainter ChromaPainter;
     public static Camera Camera = new Camera();
     public static IPlayerRenderer PlayerRenderer = new LegacyPlayerRenderer();
     public static IPlayerRenderer PotionOfReturnRenderer = new ReturnGatePlayerRenderer();
@@ -868,7 +_,7 @@
     public static HairstyleUnlocksHelper Hairstyles = new HairstyleUnlocksHelper();
     public static bool tilesLoaded = false;
     public static WorldMap Map = new WorldMap(maxTilesX, maxTilesY);
-    public static Tile[,] tile = new Tile[maxTilesX, maxTilesY];
+    public static Tilemap tile = new Tilemap(maxTilesX, maxTilesY);
     public static Dust[] dust = new Dust[6001];
     public static Star[] star = new Star[400];
     public static Item[] item = new Item[401];
@@ -970,9 +_,9 @@
     public static bool Support4K = true;
     public static float MinimumZoomComparerX = 1920f;
     public static float MinimumZoomComparerY = 1200f;
-    private static int _renderTargetMaxSize = 2048;
+    private static int _renderTargetMaxSize = 8192;
-    private static GraphicsProfile _selectedGraphicsProfile = GraphicsProfile.Reach;
+    private static GraphicsProfile _selectedGraphicsProfile = GraphicsProfile.HiDef;
-    private static GraphicsProfile _currentGraphicsProfile = GraphicsProfile.Reach;
+    private static GraphicsProfile _currentGraphicsProfile = GraphicsProfile.HiDef;
     public static int LogicCheckScreenWidth = 1920;
     public static int LogicCheckScreenHeight = 1200;
     private static WorldPreparationState _worldPreparationState = WorldPreparationState.AwaitingData;
@@ -985,9 +_,7 @@
     private static float backSpaceRate;
     public static string motd = "";
     public static bool toggleFullscreen;
-    public static int numDisplayModes;
+    //public static int numDisplayModes;
-    public static int[] displayWidth = new int[99];
-    public static int[] displayHeight = new int[99];
     public static bool gameMenu = true;
     public static bool menuBGChangedDay = false;
     public static bool menuBGChangedNight = false;
@@ -1168,7 +_,7 @@
     private float logoScale = 1f;
     private float logoScaleDirection = 1f;
     private float logoScaleSpeed = 1f;
-    private static int maxMenuItems = 16;
+    private static int maxMenuItems = 17;
     private float[] menuItemScale = new float[maxMenuItems];
     private int focusMenu = -1;
     private int selectedMenu = -1;
@@ -2425,7 +_,6 @@
         Configuration.Clear();
         Configuration.Put("ResetDefaultUIScale", false);
         Configuration.Put("SmartCursorToggle", cSmartCursorModeIsToggleAndNotHold);
-        Configuration.Put("MapEnabled", mapEnabled);
         Configuration.Put("InvasionBarMode", invasionProgressMode);
         Configuration.Put("AutoSave", autoSave);
         Configuration.Put("AutoPause", autoPause);
@@ -2578,22 +_,25 @@
 
         PendingBorderlessState = screenBorderless;
         screenBorderlessPendingResizes = (screenBorderless ? 6 : 0);
-        Form form = (Form)Control.FromHandle(instance.Window.Handle);
+
         if (screenBorderless)
         {
-            SetBorderlessFormStyle(form);
+            SetBorderlessFormStyle();
         }
         else if (screenMaximized)
         {
-            form.FormBorderStyle = FormBorderStyle.Sizable;
-            form.WindowState = FormWindowState.Maximized;
+            SDL2.SDL.SDL_SetWindowBordered(Main.instance.Window.Handle, SDL2.SDL.SDL_bool.SDL_TRUE);
+            SDL2.SDL.SDL_SetWindowResizable(Main.instance.Window.Handle, SDL2.SDL.SDL_bool.SDL_TRUE);
+            SDL2.SDL.SDL_MaximizeWindow(Main.instance.Window.Handle);
         }
         else
         {
-            form.FormBorderStyle = FormBorderStyle.Sizable;
+            SDL2.SDL.SDL_SetWindowBordered(Main.instance.Window.Handle, SDL2.SDL.SDL_bool.SDL_TRUE);
+            SDL2.SDL.SDL_SetWindowResizable(Main.instance.Window.Handle, SDL2.SDL.SDL_bool.SDL_TRUE);
         }
 
-        form.BringToFront();
+        SDL2.SDL.SDL_SetWindowInputFocus(Main.instance.Window.Handle);
+
         int currentValue2 = graphics.PreferredBackBufferWidth;
         int currentValue3 = graphics.PreferredBackBufferHeight;
         Configuration.Get("DisplayWidth", ref currentValue2);
@@ -2611,7 +_,6 @@
         }
 
         Configuration.Get("SmartCursorToggle", ref cSmartCursorModeIsToggleAndNotHold);
-        Configuration.Get("MapEnabled", ref mapEnabled);
         Configuration.Get("InvasionBarMode", ref invasionProgressMode);
         Configuration.Get("AutoSave", ref autoSave);
         Configuration.Get("AutoPause", ref autoPause);
@@ -2982,7 +_,7 @@
 
                 if (num >= 91 && binaryReader.ReadBoolean())
                 {
-                    ((Form)Control.FromHandle(base.Window.Handle)).WindowState = FormWindowState.Maximized;
+                    //((Form)Control.FromHandle(base.Window.Handle)).WindowState = FormWindowState.Maximized;
                 }
 
                 if (num >= 4)
@@ -4486,6 +_,7 @@
         }
     }
 
+    static bool didResize = false;
     public Main()
     {
         instance = this;
@@ -4494,6 +_,16 @@
         Configuration.Load();
         graphics = new GraphicsDeviceManager(this as Game);
         base.Content.RootDirectory = "Content";
+        base.Window.AllowUserResizing = true;
+        base.Window.ClientSizeChanged += (x, y) =>
+        {
+            if (!graphics.IsFullScreen && !screenBorderless)
+            {
+                SDL2.SDL.SDL_GetWindowSize(Window.Handle, out int w, out int h);
+                SetResolution(w, h);
+                didResize = true;
+            }
+        };
     }
 
     private static void SetDisplayMonitor()
@@ -8046,9 +_,6 @@
         _achievementAdvisor = new AchievementAdvisor();
         OpenRecent();
         UILinksInitializer.Load();
-        Chroma = new ChromaEngine();
-        ChromaPainter = new ChromaHotkeyPainter();
-        ChromaPainter.CollectBoundKeys();
         CacheSupportedDisplaySizes();
         if (autoJoin)
         {
@@ -8062,31 +_,35 @@
 
     private static void CacheSupportedDisplaySizes()
     {
-        numDisplayModes = 0;
-        foreach (DisplayMode supportedDisplayMode in GraphicsAdapter.DefaultAdapter.SupportedDisplayModes)
-        {
-            if (supportedDisplayMode.Width < minScreenW || supportedDisplayMode.Height < minScreenH || supportedDisplayMode.Width > maxScreenW || supportedDisplayMode.Height > maxScreenH)
-            {
-                continue;
-            }
-
-            bool flag = true;
-            for (int i = 0; i < numDisplayModes; i++)
-            {
-                if (supportedDisplayMode.Width == displayWidth[i] && supportedDisplayMode.Height == displayHeight[i])
-                {
-                    flag = false;
-                    break;
-                }
-            }
-
-            if (flag)
-            {
-                displayHeight[numDisplayModes] = supportedDisplayMode.Height;
-                displayWidth[numDisplayModes] = supportedDisplayMode.Width;
-                numDisplayModes++;
-            }
-        }
+        SDL2.SDL.SDL_GetDisplayBounds(SDL2.SDL.SDL_GetWindowDisplayIndex(Main.instance.Window.Handle), out SDL2.SDL.SDL_Rect bounds);
+
+        Main.maxScreenW = bounds.w;
+        Main.maxScreenH = bounds.h;
+        //numDisplayModes = 0;
+        //foreach (DisplayMode supportedDisplayMode in GraphicsAdapter.DefaultAdapter.SupportedDisplayModes)
+        //{
+        //    if (supportedDisplayMode.Width < minScreenW || supportedDisplayMode.Height < minScreenH || supportedDisplayMode.Width > maxScreenW || supportedDisplayMode.Height > maxScreenH)
+        //    {
+        //        continue;
+        //    }
+        //
+        //    bool flag = true;
+        //    for (int i = 0; i < numDisplayModes; i++)
+        //    {
+        //        if (supportedDisplayMode.Width == displayWidth[i] && supportedDisplayMode.Height == displayHeight[i])
+        //        {
+        //            flag = false;
+        //            break;
+        //        }
+        //    }
+        //
+        //    if (flag)
+        //    {
+        //        displayHeight[numDisplayModes] = supportedDisplayMode.Height;
+        //        displayWidth[numDisplayModes] = supportedDisplayMode.Width;
+        //        numDisplayModes++;
+        //    }
+        //}
     }
 
     public static void LoadTestLog(string logname)
@@ -8137,7 +_,6 @@
 
         AssetSourceController.UseResourcePacks(AssetInitializer.CreateResourcePackList(base.Services));
         AssetInitializer.LoadSplashAssets(asyncLoadForSounds: true);
-        ChromaInitializer.Load();
         _gameContentLoadProcess = LoadContent_Deferred();
     }
 
@@ -9867,7 +_,7 @@
             {
                 newMusic = 10;
             }
-            else
+            else if (tile[(int)(player[myPlayer].Center.X / 16f), (int)(player[myPlayer].Center.Y / 16f)] != null && tile[(int)(player[myPlayer].Center.X / 16f), (int)(player[myPlayer].Center.Y / 16f)].wall == 87)
             {
                 newMusic = 8;
             }
@@ -13876,8 +_,6 @@
         {
             UpdateAudio();
             GlobalTimeWrappedHourly = (float)(gameTime.TotalGameTime.TotalSeconds % 3600.0);
-            ChromaInitializer.UpdateEvents();
-            Chroma.Update(GlobalTimeWrappedHourly);
             return;
         }
 
@@ -14011,8 +_,6 @@
         DoUpdate_AutoSave();
         if (!dedServ)
         {
-            ChromaInitializer.UpdateEvents();
-            Chroma.Update(GlobalTimeWrappedHourly);
             if (superFast)
             {
                 base.IsFixedTimeStep = false;
@@ -14222,14 +_,6 @@
 
             UpdateOldNPCShop();
             hasFocus = base.IsActive;
-            Form form = Control.FromHandle(base.Window.Handle) as Form;
-            bool num2 = form.WindowState == FormWindowState.Minimized;
-            bool flag = Form.ActiveForm == form;
-            hasFocus |= flag;
-            if (num2)
-            {
-                hasFocus = false;
-            }
 
             if (!hasFocus && netMode == 0)
             {
@@ -14391,10 +_,8 @@
         {
             DoUpdateInWorld(_worldUpdateTimeTester);
         }
-
         if (netMode != 2)
         {
-            ChromaPainter.Update();
         }
     }
 
@@ -15111,11 +_,6 @@
             {
                 UpdateServer();
             }
-
-            if (netMode == 1)
-            {
-                UpdateClient();
-            }
         }
 
         chatMonitor.Update();
@@ -17801,7 +_,7 @@
                     {
                         toolTipLine[numLines] = Lang.tip[13].Value;
                     }
-
+                    toolTipLine[numLines] += $" [{item.useAnimation}/{(item.useAnimation / 60f):F2}s{(item.useAnimation != item.useTime ? $" {item.useTime}/{(item.useTime / 60f):F2}s" : "")}]";
                     numLines++;
                 }
 
@@ -17858,6 +_,8 @@
                     toolTipLine[numLines] = Lang.tip[22].Value;
                 }
 
+                toolTipLine[numLines] += $" [{num6:F2}]";
+
                 numLines++;
             }
 
@@ -18344,6 +_,7 @@
                     toolTipLine[numLines] = "+4" + Lang.tip[47].Value;
                     preFixLine[numLines] = true;
                     numLines++;
+
                 }
             }
 
@@ -43835,7 +_,6 @@
 
     private static void TryDisposingEverything()
     {
-        ChromaInitializer.DisableAllDeviceGroups();
         CaptureManager.Instance.Dispose();
         audioSystem.Dispose();
     }
@@ -44961,6 +_,26 @@
             }
 
             num11++;
+            num5++;
+            array9[num11] = "TerraAngel Settings";
+            if (selectedMenu == num11)
+            {
+                SoundEngine.PlaySound(10);
+                menuMode = 888;
+                MenuUI.SetState(ClientLoader.ConfigUI);
+            }
+
+            num11++;
+            num5++;
+            array9[num11] = "TerraAngel Plugins";
+            if (selectedMenu == num11)
+            {
+                SoundEngine.PlaySound(10);
+                menuMode = 888;
+                MenuUI.SetState(ClientLoader.PluginUI);
+            }
+
+            num11++;
             array9[num11] = Lang.menu[14].Value;
             if (selectedMenu == num11)
             {
@@ -45735,6 +_,8 @@
                     menuMode = 112;
                 }
 
+
+
                 num20++;
                 array9[num20] = Lang.menu[210].Value;
                 if (selectedMenu == num20)
@@ -46161,83 +_,87 @@
             }
             else if (menuMode == 111)
             {
-                for (int num29 = 0; num29 < 9; num29++)
-                {
-                    array7[num29] = 0.85f;
-                }
-
-                num2 = 210;
-                num4 = 55;
-                int num30 = 0;
-                array9[num30] = Lang.menu[73].Value + ": " + PendingResolutionWidth + "x" + PendingResolutionHeight;
-                if (selectedMenu == num30)
-                {
-                    SoundEngine.PlaySound(12);
-                    int num31 = 0;
-                    for (int num32 = 0; num32 < numDisplayModes; num32++)
-                    {
-                        if (displayWidth[num32] == PendingResolutionWidth && displayHeight[num32] == PendingResolutionHeight)
-                        {
-                            num31 = num32;
-                            break;
-                        }
-                    }
-
-                    num31 = (num31 + 1) % numDisplayModes;
-                    PendingResolutionWidth = displayWidth[num31];
-                    PendingResolutionHeight = displayHeight[num31];
-                }
-
-                num30++;
-                if (IsBorderlessDisplayAvailable())
-                {
-                    array9[num30] = Lang.menu[PendingBorderlessState ? 245 : 246].Value;
-                    if (selectedMenu == num30)
-                    {
-                        SoundEngine.PlaySound(12);
-                        PendingBorderlessState = !PendingBorderlessState;
-                    }
-
-                    num30++;
-                }
-
-                array9[num30] = (graphics.IsFullScreen ? Lang.menu[49].Value : Lang.menu[50].Value);
-                if (selectedMenu == num30)
-                {
-                    ToggleFullScreen();
-                }
-
-                num30++;
-                array4[num30] = 100;
-                array9[num30] = Lang.menu[134].Value;
-                if (selectedMenu == num30)
-                {
-                    if (graphics.IsFullScreen || PendingBorderlessState != screenBorderless || PendingResolutionWidth != screenWidth || PendingResolutionHeight != screenHeight)
-                    {
-                        screenBorderless = PendingBorderlessState;
-                        screenBorderlessPendingResizes = (screenBorderless ? 6 : 0);
-                        SetResolution(PendingResolutionWidth, PendingResolutionHeight);
-                    }
-
-                    SoundEngine.PlaySound(11);
-                    menuMode = 1111;
-                }
-
-                num30++;
-                array9[num30] = Lang.menu[5].Value;
-                array4[num30] = 100;
-                if (selectedMenu == num30 || flag5)
-                {
-                    flag5 = false;
-                    PendingResolutionWidth = graphics.PreferredBackBufferWidth;
-                    PendingResolutionHeight = graphics.PreferredBackBufferHeight;
-                    PendingBorderlessState = screenBorderless;
-                    menuMode = 1111;
-                    SoundEngine.PlaySound(11);
-                }
-
-                num30++;
-                num5 = num30;
+                SoundEngine.PlaySound(10);
+                menuMode = 888;
+                MenuUI.SetState(ClientLoader.ResolutionUI);
+
+                //for (int num29 = 0; num29 < 9; num29++)
+                //{
+                //    array7[num29] = 0.85f;
+                //}
+
+                //num2 = 210;
+                //num4 = 55;
+                //int num30 = 0;
+                //array9[num30] = Lang.menu[73].Value + ": " + PendingResolutionWidth + "x" + PendingResolutionHeight;
+                //if (selectedMenu == num30)
+                //{
+                //    SoundEngine.PlaySound(12);
+                //    int num31 = 0;
+                //    for (int num32 = 0; num32 < numDisplayModes; num32++)
+                //    {
+                //        if (displayWidth[num32] == PendingResolutionWidth && displayHeight[num32] == PendingResolutionHeight)
+                //        {
+                //            num31 = num32;
+                //            break;
+                //        }
+                //    }
+
+                //    num31 = (num31 + 1) % numDisplayModes;
+                //    PendingResolutionWidth = displayWidth[num31];
+                //    PendingResolutionHeight = displayHeight[num31];
+                //}
+
+                //num30++;
+                //if (IsBorderlessDisplayAvailable())
+                //{
+                //    array9[num30] = Lang.menu[PendingBorderlessState ? 245 : 246].Value;
+                //    if (selectedMenu == num30)
+                //    {
+                //        SoundEngine.PlaySound(12);
+                //        PendingBorderlessState = !PendingBorderlessState;
+                //    }
+
+                //    num30++;
+                //}
+
+                //array9[num30] = (graphics.IsFullScreen ? Lang.menu[49].Value : Lang.menu[50].Value);
+                //if (selectedMenu == num30)
+                //{
+                //    ToggleFullScreen();
+                //}
+
+                //num30++;
+                //array4[num30] = 100;
+                //array9[num30] = Lang.menu[134].Value;
+                //if (selectedMenu == num30)
+                //{
+                //    if (graphics.IsFullScreen || PendingBorderlessState != screenBorderless || PendingResolutionWidth != screenWidth || PendingResolutionHeight != screenHeight)
+                //    {
+                //        screenBorderless = PendingBorderlessState;
+                //        screenBorderlessPendingResizes = (screenBorderless ? 6 : 0);
+                //        SetResolution(PendingResolutionWidth, PendingResolutionHeight);
+                //    }
+
+                //    SoundEngine.PlaySound(11);
+                //    menuMode = 1111;
+                //}
+
+                //num30++;
+                //array9[num30] = Lang.menu[5].Value;
+                //array4[num30] = 100;
+                //if (selectedMenu == num30 || flag5)
+                //{
+                //    flag5 = false;
+                //    PendingResolutionWidth = graphics.PreferredBackBufferWidth;
+                //    PendingResolutionHeight = graphics.PreferredBackBufferHeight;
+                //    PendingBorderlessState = screenBorderless;
+                //    menuMode = 1111;
+                //    SoundEngine.PlaySound(11);
+                //}
+
+                //num30++;
+                //num5 = num30;
             }
             else if (menuMode == 1125)
             {
@@ -48405,35 +_,40 @@
         return true;
     }
 
-    private static void SetDisplayModeAsBorderless(ref int width, ref int height, Form form)
+    public static void SetDisplayModeAsBorderless(ref int width, ref int height)
     {
         if (screenBorderless && !graphics.IsFullScreen && screenBorderlessPendingResizes > 0)
         {
             screenBorderlessPendingResizes--;
-            System.Drawing.Rectangle bounds = Screen.FromPoint(form.Location).Bounds;
+            SDL2.SDL.SDL_GetDisplayBounds(SDL2.SDL.SDL_GetWindowDisplayIndex(Main.instance.Window.Handle), out SDL2.SDL.SDL_Rect bounds);
-            width = bounds.Width;
+            width = bounds.w;
-            height = bounds.Height;
+            height = bounds.h;
+            ApplyBorderlessResolution();
             TryPickingDefaultUIScale(height);
         }
     }
 
-    private static void ApplyBorderlessResolution(Form form)
+    private static void ApplyBorderlessResolution()
     {
         if (screenBorderlessPendingResizes > 0)
         {
             screenBorderlessPendingResizes--;
-            System.Drawing.Rectangle bounds = Screen.FromPoint(form.Location).Bounds;
-            form.Location = new System.Drawing.Point(bounds.X, bounds.Y);
-            form.FormBorderStyle = FormBorderStyle.None;
-            form.Width = bounds.Width;
-            form.Height = bounds.Height;
+
+            SDL2.SDL.SDL_GetDisplayBounds(SDL2.SDL.SDL_GetWindowDisplayIndex(Main.instance.Window.Handle), out SDL2.SDL.SDL_Rect bounds);
+            SDL2.SDL.SDL_SetWindowBordered(Main.instance.Window.Handle, SDL2.SDL.SDL_bool.SDL_FALSE);
+            SDL2.SDL.SDL_SetWindowPosition(Main.instance.Window.Handle, bounds.x, bounds.y);
+            SDL2.SDL.SDL_SetWindowSize(Main.instance.Window.Handle, bounds.w, bounds.h);
+
+            Main.screenWidth = bounds.w;
+            Main.screenHeight = bounds.h;
         }
     }
 
-    private static void SetBorderlessFormStyle(Form form)
+    private static void SetBorderlessFormStyle()
     {
-        form.Location = new System.Drawing.Point(0, 0);
-        form.FormBorderStyle = FormBorderStyle.None;
+        SDL2.SDL.SDL_SetWindowPosition(Main.instance.Window.Handle, 0, 0);
+        SDL2.SDL.SDL_SetWindowBordered(Main.instance.Window.Handle, SDL2.SDL.SDL_bool.SDL_FALSE);
+        SDL2.SDL.SDL_SetWindowResizable(Main.instance.Window.Handle, SDL2.SDL.SDL_bool.SDL_FALSE);
     }
 
     public static void OpenCharacterSelectUI()
@@ -52932,6 +_,10 @@
 
     private static void DrawMapFullscreenBackground(Vector2 screenPosition, int screenWidth, int screenHeight)
     {
+        if (!WorldGen.InWorld((int)(player[myPlayer].Center.X / 16f), (int)(player[myPlayer].Center.Y / 16f)) || tile[(int)(player[myPlayer].Center.X / 16f), (int)(player[myPlayer].Center.Y / 16f)] == null)
+        {
+            return;
+        }
         Asset<Texture2D> asset = TextureAssets.MapBGs[0];
         int num = -1;
         Microsoft.Xna.Framework.Color color = Microsoft.Xna.Framework.Color.White;
@@ -58862,7 +_,13 @@
                 DrawWires();
             }
 
-            TimeLogger.DetailedDrawTime(26);
+            if (!Main.didResize)
+            {
+                SDL2.SDL.SDL_GetWindowSize(instance.Window.Handle, out int w, out int h);
+                base.GraphicsDevice.Viewport = new Viewport(0, 0, w, h);
+                SetResolution(base.GraphicsDevice.Viewport.Width, base.GraphicsDevice.Viewport.Height);
+            }
+            Main.didResize = false;
         }
 
         Overlays.Scene.Draw(spriteBatch, RenderLayers.ForegroundWater);
@@ -59441,8 +_,6 @@
         }
 
         instance.CameraModifiers.ApplyTo(ref screenPosition);
-        screenPosition.X = (int)screenPosition.X;
-        screenPosition.Y = (int)screenPosition.Y;
         ClampScreenPositionToWorld();
     }
 
@@ -62940,15 +_,19 @@
     public static void SetDisplayMode(int width, int height, bool fullscreen)
     {
         bool flag = false;
-        Form form = (Form)Control.FromHandle(instance.Window.Handle);
-        screenMaximized = form.WindowState == FormWindowState.Maximized;
+        IntPtr window = Main.instance.Window.Handle;
+        uint flags = SDL2.SDL.SDL_GetWindowFlags(window);
+
+        screenMaximized = (flags & (uint)SDL2.SDL.SDL_WindowFlags.SDL_WINDOW_MAXIMIZED) != 0;
+
         if (screenBorderless && screenMaximized && !graphics.IsFullScreen)
         {
             screenMaximized = false;
-            form.WindowState = FormWindowState.Normal;
+            SDL2.SDL.SDL_RestoreWindow(window);
         }
 
-        flag = form.FormBorderStyle == FormBorderStyle.None;
+        flag = (flags & (uint)SDL2.SDL.SDL_WindowFlags.SDL_WINDOW_BORDERLESS) != 0;
+
         bool flag2 = false;
         int num3;
         int num4;
@@ -62966,10 +_,10 @@
                 }
             }
 
-            form.MinimumSize = new Size(0, 0);
+            SDL2.SDL.SDL_SetWindowMinimumSize(window, 0, 0);
             if (!fullscreen && !flag3)
             {
-                SetDisplayModeAsBorderless(ref width, ref height, form);
+                SetDisplayModeAsBorderless(ref width, ref height);
             }
 
             if (width > maxScreenW)
@@ -63005,11 +_,12 @@
         else
         {
             PlayerInput.RawMouseScale = Vector2.One;
-            form.MinimumSize = new Size(minScreenW, minScreenH);
+
+            SDL2.SDL.SDL_SetWindowMinimumSize(window, minScreenW, minScreenH);
             if (flag)
             {
-                width = displayWidth[0];
-                height = displayHeight[0];
+                width = 800;
+                height = 600;
             }
 
             width = Math.Min(width, maxScreenW);
@@ -63021,21 +_,31 @@
 
         if (!fullscreen && !flag2)
         {
+            SDL2.SDL.SDL_GetWindowSize(window, out int w, out int h);
-            if (form.ClientSize.Width < graphics.PreferredBackBufferWidth)
+            if (w < graphics.PreferredBackBufferWidth)
             {
-                width = form.ClientSize.Width;
-                flag2 = true;
+                width = w;
             }
-
-            if (form.ClientSize.Height < graphics.PreferredBackBufferHeight)
+            if (h < graphics.PreferredBackBufferHeight)
             {
-                height = form.ClientSize.Height;
-                flag2 = true;
+                height = h;
             }
         }
 
-        width &= 0x7FFFFFFE;
+        //width &= 0x7FFFFFFE;
-        height &= 0x7FFFFFFE;
+        //height &= 0x7FFFFFFE;
+
+        if (screenBorderless && !flag)
+        {
+            SDL2.SDL.SDL_GetDisplayBounds(SDL2.SDL.SDL_GetWindowDisplayIndex(Main.instance.Window.Handle), out SDL2.SDL.SDL_Rect bounds);
+            SDL2.SDL.SDL_SetWindowBordered(Main.instance.Window.Handle, SDL2.SDL.SDL_bool.SDL_FALSE);
+            SDL2.SDL.SDL_SetWindowPosition(Main.instance.Window.Handle, bounds.x, bounds.y);
+            SDL2.SDL.SDL_SetWindowSize(Main.instance.Window.Handle, bounds.w, bounds.h);
+
+            Main.screenWidth = bounds.w;
+            Main.screenHeight = bounds.h;
+        }
+
         width = Math.Max(width, minScreenW);
         height = Math.Max(height, minScreenH);
         if (graphics.IsFullScreen != fullscreen)
@@ -63045,7 +_,21 @@
             graphics.ApplyChanges();
             graphics.ToggleFullScreen();
         }
+        SDL2.SDL.SDL_GetWindowSize(window, out int ww, out int wh);
 
+        if (graphics.PreferredBackBufferWidth != ww)
+        {
+            graphics.PreferredBackBufferWidth = ww;
+        }
+        if (graphics.PreferredBackBufferHeight != wh)
+        {
+            graphics.PreferredBackBufferHeight = wh;
+        }
+
+        if (graphics.GraphicsDevice.Viewport.Width != ww || graphics.GraphicsDevice.Viewport.Height != wh)
+        {
+            graphics.GraphicsDevice.Viewport = new Viewport(0, 0, ww, wh);
+        }
         if (width != num3 || height != num4 || flag2)
         {
             mapTime = 0;
@@ -63073,18 +_,15 @@
             {
                 if (screenBorderless)
                 {
-                    ApplyBorderlessResolution(form);
+                    ApplyBorderlessResolution();
-                    form.FormBorderStyle = FormBorderStyle.None;
                 }
                 else
                 {
-                    form.FormBorderStyle = FormBorderStyle.Sizable;
+                    SDL2.SDL.SDL_SetWindowResizable(window, SDL2.SDL.SDL_bool.SDL_TRUE);
+                    SDL2.SDL.SDL_SetWindowBordered(window, SDL2.SDL.SDL_bool.SDL_TRUE);
                 }
-
-                form.SendToBack();
-                form.BringToFront();
+                SDL2.SDL.SDL_SetWindowInputFocus(window);
             }
-
             Lighting.Initialize();
             if (!drawToScreen && !_isResizingAndRemakingTargets)
             {
@@ -63128,7 +_,13 @@
             FullscreenStartup();
         }
 
+        if (!Main.didResize)
+        {
+            SDL2.SDL.SDL_GetWindowSize(instance.Window.Handle, out int w, out int h);
+            base.GraphicsDevice.Viewport = new Viewport(0, 0, w, h);
-        SetResolution(base.GraphicsDevice.Viewport.Width, base.GraphicsDevice.Viewport.Height);
+            SetResolution(base.GraphicsDevice.Viewport.Width, base.GraphicsDevice.Viewport.Height);
+        }
+        Main.didResize = false;
     }
 
     public static void OpenPlayerSelect(OnPlayerSelected method)
