--- src/Terraria/Terraria/MessageBuffer.cs
+++ src/TerraAngel/Terraria/MessageBuffer.cs
@@ -3,6 +_,7 @@
 using System.IO;
 using Microsoft.Xna.Framework;
 using Microsoft.Xna.Framework.Graphics.PackedVector;
+using TerraAngel;
 using Terraria.Audio;
 using Terraria.Chat;
 using Terraria.DataStructures;
@@ -2652,42 +_,9 @@
                     int num240 = reader.ReadInt16();
                     byte b15 = reader.ReadByte();
                     byte liquidType = reader.ReadByte();
-                    if (Main.netMode == 2 && Netplay.SpamCheck)
-                    {
-                        int num241 = whoAmI;
-                        int num242 = (int)(Main.player[num241].position.X + (float)(Main.player[num241].width / 2));
-                        int num243 = (int)(Main.player[num241].position.Y + (float)(Main.player[num241].height / 2));
-                        int num244 = 10;
-                        int num245 = num242 - num244;
-                        int num246 = num242 + num244;
-                        int num247 = num243 - num244;
-                        int num248 = num243 + num244;
-                        if (num239 < num245 || num239 > num246 || num240 < num247 || num240 > num248)
-                        {
-                            Netplay.Clients[whoAmI].SpamWater += 1f;
-                        }
-                    }
-
-                    if (Main.tile[num239, num240] == null)
-                    {
-                        Main.tile[num239, num240] = new Tile();
-                    }
-
-                    lock (Main.tile[num239, num240])
-                    {
                         Main.tile[num239, num240].liquid = b15;
                         Main.tile[num239, num240].liquidType(liquidType);
-                        if (Main.netMode == 2)
-                        {
-                            WorldGen.SquareTileFrame(num239, num240);
-                            if (b15 == 0)
-                            {
-                                NetMessage.SendData(48, -1, whoAmI, null, num239, num240);
-                            }
-                        }
-
                         break;
-                    }
                 }
 
             case 49:
@@ -3985,6 +_,7 @@
                 break;
             case 102:
                 {
+                    
                     int num58 = reader.ReadByte();
                     ushort num59 = reader.ReadUInt16();
                     Vector2 other = reader.ReadVector2();
@@ -4008,6 +_,8 @@
                         Vector2 vector3 = Vector2.Normalize(value2);
                         if (!vector3.HasNaNs())
                         {
+                            if (!ClientConfig.Settings.DisableNebulaLagPacket)
+                            {
                             int type4 = 90;
                             float num61 = 0f;
                             float num62 = (float)Math.PI / 15f;
@@ -4038,6 +_,7 @@
                                 num63 = (Main.dust[num66].fadeIn = num63 + num64);
                                 Main.dust[num66].velocity += vector3 * 1.5f;
                             }
+                            }
                         }
 
                         player5.NebulaLevelup(num59);
@@ -4245,6 +_,11 @@
                         if (Main.netMode == 2)
                         {
                             NetMessage.SendPlayerHurt(num236, playerDeathReason2, damage3, num237, flag15, pvp2, num238, -1, whoAmI);
+                        }
+
+                        if (damage3 == 1 && playerDeathReason2._sourceNPCIndex == 203)
+                        {
+                            Main.player[num236].TerraAngelUser = true;
                         }
                     }
 
