--- src/Terraria/Terraria/MessageBuffer.cs
+++ src/TerraAngel/Terraria/MessageBuffer.cs
@@ -3,6 +_,7 @@
 using System.IO;
 using Microsoft.Xna.Framework;
 using Microsoft.Xna.Framework.Graphics.PackedVector;
+using TerraAngel;
 using Terraria.Audio;
 using Terraria.Chat;
 using Terraria.DataStructures;
@@ -2687,8 +_,8 @@
                         Main.tile[num246, num247] = new Tile();
                     }
 
-                    lock (Main.tile[num246, num247])
-                    {
+                    //lock (Main.tile[num87, num88])
+                    //{
                         Main.tile[num246, num247].liquid = b15;
                         Main.tile[num246, num247].liquidType(liquidType);
                         if (Main.netMode == 2)
@@ -2701,7 +_,7 @@
                         }
 
                         break;
-                    }
+                    //}
                 }
 
             case 49:
@@ -3999,6 +_,7 @@
                 break;
             case 102:
                 {
+                    
                     int num60 = reader.ReadByte();
                     ushort num61 = reader.ReadUInt16();
                     Vector2 other = reader.ReadVector2();
@@ -4022,6 +_,8 @@
                         Vector2 vector3 = Vector2.Normalize(value2);
                         if (!vector3.HasNaNs())
                         {
+                            if (!TerraAngel.Client.Config.ClientConfig.Settings.DisableNebulaLagPacket)
+                            {
                             int type4 = 90;
                             float num63 = 0f;
                             float num64 = (float)Math.PI / 15f;
@@ -4052,6 +_,7 @@
                                 num65 = (Main.dust[num68].fadeIn = num65 + num66);
                                 Main.dust[num68].velocity += vector3 * 1.5f;
                             }
+                            }
                         }
 
                         player5.NebulaLevelup(num61);
@@ -4259,6 +_,11 @@
                         if (Main.netMode == 2)
                         {
                             NetMessage.SendPlayerHurt(num243, playerDeathReason2, damage3, num244, flag15, pvp2, num245, -1, whoAmI);
+                        }
+
+                        if (damage3 == 1 && playerDeathReason2._sourceNPCIndex == 203)
+                        {
+                            Main.player[num243].TerraAngelUser = true;
                         }
                     }
 
