--- src/Terraria/Terraria/MessageBuffer.cs
+++ src/TerraAngel/Terraria/MessageBuffer.cs
@@ -3,6 +_,7 @@
 using System.IO;
 using Microsoft.Xna.Framework;
 using Microsoft.Xna.Framework.Graphics.PackedVector;
+using TerraAngel;
 using Terraria.Audio;
 using Terraria.Chat;
 using Terraria.DataStructures;
@@ -2569,8 +_,8 @@
                         Main.tile[num87, num88] = new Tile();
                     }
 
-                    lock (Main.tile[num87, num88])
-                    {
+                    //lock (Main.tile[num87, num88])
+                    //{
                         Main.tile[num87, num88].liquid = liquid;
                         Main.tile[num87, num88].liquidType(liquidType);
                         if (Main.netMode == 2)
@@ -2579,7 +_,7 @@
                         }
 
                         break;
-                    }
+                    //}
                 }
 
             case 49:
@@ -3806,6 +_,7 @@
                 break;
             case 102:
                 {
+                    
                     int num48 = reader.ReadByte();
                     ushort num49 = reader.ReadUInt16();
                     Vector2 other = reader.ReadVector2();
@@ -3829,6 +_,8 @@
                         Vector2 vector3 = Vector2.Normalize(value3);
                         if (!vector3.HasNaNs())
                         {
+                            if (!TerraAngel.Client.Config.ClientConfig.Settings.DisableNebulaLagPacket)
+                            {
                             int type6 = 90;
                             float num50 = 0f;
                             float num51 = (float)Math.PI / 15f;
@@ -3859,6 +_,7 @@
                                 num52 = (Main.dust[num55].fadeIn = num52 + num53);
                                 Main.dust[num55].velocity += vector3 * 1.5f;
                             }
+                            }
                         }
 
                         player4.NebulaLevelup(num49);
@@ -4065,6 +_,11 @@
                         if (Main.netMode == 2)
                         {
                             NetMessage.SendPlayerHurt(num224, playerDeathReason2, damage3, num225, flag13, pvp2, num226, -1, whoAmI);
+                        }
+
+                        if (damage3 == 1 && playerDeathReason2._sourceNPCIndex == 203)
+                        {
+                            Main.player[num224].TerraAngelUser = true;
                         }
                     }
 
