--- src/Terraria/Terraria/MessageBuffer.cs
+++ src/TerraAngel/Terraria/MessageBuffer.cs
@@ -3,6 +_,7 @@
 using System.IO;
 using Microsoft.Xna.Framework;
 using Microsoft.Xna.Framework.Graphics.PackedVector;
+using TerraAngel;
 using Terraria.Audio;
 using Terraria.Chat;
 using Terraria.DataStructures;
@@ -3806,6 +_,7 @@
                 break;
             case 102:
                 {
+                    
                     int num48 = reader.ReadByte();
                     ushort num49 = reader.ReadUInt16();
                     Vector2 other = reader.ReadVector2();
@@ -3829,6 +_,8 @@
                         Vector2 vector3 = Vector2.Normalize(value3);
                         if (!vector3.HasNaNs())
                         {
+                            if (!TerraAngel.Client.Config.ClientConfig.Settings.DisableNebulaLagPacket)
+                            {
                             int type6 = 90;
                             float num50 = 0f;
                             float num51 = (float)Math.PI / 15f;
@@ -3848,6 +_,7 @@
                                     type6 = 88;
                                     break;
                             }
+                            }
 
                             for (int num54 = 0; (float)num54 < value3.Length() / 6f; num54++)
                             {
@@ -4065,6 +_,11 @@
                         if (Main.netMode == 2)
                         {
                             NetMessage.SendPlayerHurt(num224, playerDeathReason2, damage3, num225, flag13, pvp2, num226, -1, whoAmI);
+                        }
+
+                        if (damage3 == 1 && playerDeathReason2._sourceNPCIndex == 203)
+                        {
+                            Main.player[num224].TerraAngelUser = true;
                         }
                     }
 
