--- src/Terraria/Terraria/NetMessage.cs
+++ src/TerraAngel/Terraria/NetMessage.cs
@@ -2530,23 +_,26 @@
         {
             lock (buffer[i])
             {
-                try
+                if (buffer[i].totalData + streamLength < buffer[i].readBuffer.Length)
                 {
                     Buffer.BlockCopy(bytes, 0, buffer[i].readBuffer, buffer[i].totalData, streamLength);
                     buffer[i].totalData += streamLength;
                     buffer[i].checkBytes = true;
                 }
-                catch
+                else
                 {
-                    if (Main.netMode == 1)
-                    {
-                        Main.menuMode = 15;
-                        Main.statusText = Language.GetTextValue("Error.BadHeaderBufferOverflow");
-                        Netplay.Disconnect = true;
-                    }
-                    else
-                    {
-                        Netplay.Clients[i].PendingTermination = true;
+                    // Executing goes here when it overflows
+                    // Maybe log it or something
+
+                    // Clear the read buffer up a bit
+                    while (buffer[i].totalData > 200)
+                        CheckBytes(i);
+
+                    if (buffer[i].totalData + streamLength < buffer[i].readBuffer.Length)
+                    {
+                        Buffer.BlockCopy(bytes, 0, buffer[i].readBuffer, buffer[i].totalData, streamLength);
+                        buffer[i].totalData += streamLength;
+                        buffer[i].checkBytes = true;
                     }
                 }
             }
