--- src/Terraria/Terraria/NPC.cs
+++ src/TerraAngel/Terraria/NPC.cs
@@ -14,7 +_,6 @@
 using Terraria.GameContent.Drawing;
 using Terraria.GameContent.Events;
 using Terraria.GameContent.ItemDropRules;
-using Terraria.GameContent.RGB;
 using Terraria.GameContent.Tile_Entities;
 using Terraria.GameContent.UI;
 using Terraria.Graphics.CameraModifiers;
@@ -421,7 +_,7 @@
         }
     }
 
-    public string TypeName => Lang.GetNPCNameValue(netID);
+    public string TypeName => Lang.GetNPCNameValue(netID) + (ClientConfig.Settings.ShowDetailedTooltips ? $" [{type}{(Util.NPCFields.ContainsKey(type) ? ("/" + Util.NPCFields[type].Name) : "")}{(netID != type ? ("/" + netID) : "")}]" : ""); // kind of aids but shhhhhh
     public string FullName
     {
         get
@@ -1133,7 +_,6 @@
             highestTierBossOrEvent = 398;
         }
 
-        CommonConditions.Boss.HighestTierBossOrEvent = highestTierBossOrEvent;
     }
 
     public void SpawnWithHigherTime(int timeMult)
@@ -96589,6 +_,7 @@
 
     private void UpdateCollision()
     {
+        if (!Main.tile.IsTileInLoadedSection(((int)(position.X / 16f)), ((int)(position.Y / 16f)))) return;
         Collision_WalkDownSlopes();
         bool lava = Collision_LavaCollision();
         lava = Collision_WaterCollision(lava);
