--- src/Terraria/Terraria/Player.cs
+++ src/TerraAngel/Terraria/Player.cs
@@ -18171,7 +_,7 @@
 
         int num = (int)(position.X + (float)(width / 2)) / 16;
         int num2 = (int)(position.Y - 8f) / 16;
-        if (Main.tile[num, num2] == null || !Main.tile[num, num2].active() || !Main.tileRope[Main.tile[num, num2].type])
+        if (!WorldGen.InWorld(num, num2) || Main.tile[num, num2] == null || !Main.tile[num, num2].active() || !Main.tileRope[Main.tile[num, num2].type])
         {
             return;
         }
@@ -26485,216 +_,219 @@
             CheckCrackedBrickBreak();
         }
 
-        SlopeDownMovement();
-        bool flag26 = mount.Type == 7 || mount.Type == 8 || mount.Type == 12 || mount.Type == 44 || mount.Type == 49;
-        if (velocity.Y == gravity && (!mount.Active || (!mount.Cart && mount.Type != 48 && !flag26)))
+        if (!TerraAngel.Cheat.CringeManager.GetCringe<TerraAngel.Cheat.Cringes.NoClipCringe>().Enabled || i != Main.myPlayer)
         {
-            Collision.StepDown(ref position, ref velocity, width, height, ref stepSpeed, ref gfxOffY, (int)gravDir, waterWalk || waterWalk2);
-        }
+            SlopeDownMovement();
+            bool flag26 = mount.Type == 7 || mount.Type == 8 || mount.Type == 12 || mount.Type == 44 || mount.Type == 49;
+            if (velocity.Y == gravity && (!mount.Active || (!mount.Cart && mount.Type != 48 && !flag26)))
+            {
+                Collision.StepDown(ref position, ref velocity, width, height, ref stepSpeed, ref gfxOffY, (int)gravDir, waterWalk || waterWalk2);
+            }
 
-        if (gravDir == -1f)
+            if (gravDir == -1f)
-        {
+            {
-            if ((carpetFrame != -1 || velocity.Y <= gravity) && !controlUp)
+                if ((carpetFrame != -1 || velocity.Y <= gravity) && !controlUp)
+                {
+                    Collision.StepUp(ref position, ref velocity, width, height, ref stepSpeed, ref gfxOffY, (int)gravDir, controlUp);
+                }
+            }
+            else if ((carpetFrame != -1 || velocity.Y >= gravity) && !controlDown && !mount.Cart && !flag26 && grappling[0] == -1)
             {
                 Collision.StepUp(ref position, ref velocity, width, height, ref stepSpeed, ref gfxOffY, (int)gravDir, controlUp);
             }
-        }
-        else if ((carpetFrame != -1 || velocity.Y >= gravity) && !controlDown && !mount.Cart && !flag26 && grappling[0] == -1)
-        {
-            Collision.StepUp(ref position, ref velocity, width, height, ref stepSpeed, ref gfxOffY, (int)gravDir, controlUp);
-        }
 
-        oldPosition = position;
+            oldPosition = position;
-        oldDirection = direction;
+            oldDirection = direction;
-        bool falling = false;
+            bool falling = false;
-        if (velocity.Y > gravity)
+            if (velocity.Y > gravity)
-        {
+            {
-            falling = true;
+                falling = true;
-        }
+            }
 
-        if (velocity.Y < 0f - gravity)
+            if (velocity.Y < 0f - gravity)
-        {
+            {
-            falling = true;
+                falling = true;
-        }
+            }
 
-        Vector2 vector4 = velocity;
+            Vector2 vector4 = velocity;
-        slideDir = 0;
+            slideDir = 0;
-        bool ignorePlats = false;
+            bool ignorePlats = false;
-        bool fallThrough = controlDown;
+            bool fallThrough = controlDown;
-        if ((gravDir == -1f) | (mount.Active && (mount.Cart || mount.Type == 12 || mount.Type == 7 || mount.Type == 8 || mount.Type == 23 || mount.Type == 44 || mount.Type == 48)) | GoingDownWithGrapple)
+            if ((gravDir == -1f) | (mount.Active && (mount.Cart || mount.Type == 12 || mount.Type == 7 || mount.Type == 8 || mount.Type == 23 || mount.Type == 44 || mount.Type == 48)) | GoingDownWithGrapple)
-        {
+            {
-            ignorePlats = true;
+                ignorePlats = true;
-            fallThrough = true;
+                fallThrough = true;
-        }
+            }
 
-        onTrack = false;
+            onTrack = false;
-        bool flag27 = false;
+            bool flag27 = false;
-        if (mount.Active && mount.Cart)
+            if (mount.Active && mount.Cart)
-        {
+            {
-            float num100 = ((ignoreWater || merman) ? 1f : (honeyWet ? 0.25f : ((!wet) ? 1f : 0.5f)));
+                float num100 = ((ignoreWater || merman) ? 1f : (honeyWet ? 0.25f : ((!wet) ? 1f : 0.5f)));
-            velocity *= num100;
+                velocity *= num100;
-            DelegateMethods.Minecart.rotation = fullRotation;
+                DelegateMethods.Minecart.rotation = fullRotation;
-            DelegateMethods.Minecart.rotationOrigin = fullRotationOrigin;
+                DelegateMethods.Minecart.rotationOrigin = fullRotationOrigin;
-            BitsByte bitsByte = Minecart.TrackCollision(ref position, ref velocity, ref lastBoost, width, height, controlDown, controlUp, fallStart2, trackOnly: false, mount.Delegations);
+                BitsByte bitsByte = Minecart.TrackCollision(ref position, ref velocity, ref lastBoost, width, height, controlDown, controlUp, fallStart2, trackOnly: false, mount.Delegations);
-            if (bitsByte[0])
+                if (bitsByte[0])
-            {
+                {
-                onTrack = true;
+                    onTrack = true;
-                gfxOffY = Minecart.TrackRotation(ref fullRotation, position + velocity, width, height, controlDown, controlUp, mount.Delegations);
+                    gfxOffY = Minecart.TrackRotation(ref fullRotation, position + velocity, width, height, controlDown, controlUp, mount.Delegations);
-                fullRotationOrigin = new Vector2(width / 2, height);
+                    fullRotationOrigin = new Vector2(width / 2, height);
-            }
+                }
 
-            if (bitsByte[1])
+                if (bitsByte[1])
-            {
+                {
-                if (controlLeft || controlRight)
+                    if (controlLeft || controlRight)
-                {
+                    {
-                    if (cartFlip)
+                        if (cartFlip)
-                    {
+                        {
-                        cartFlip = false;
+                            cartFlip = false;
-                    }
+                        }
-                    else
+                        else
-                    {
+                        {
-                        cartFlip = true;
+                            cartFlip = true;
+                        }
                     }
-                }
 
-                if (velocity.X > 0f)
+                    if (velocity.X > 0f)
-                {
+                    {
-                    direction = 1;
+                        direction = 1;
-                }
+                    }
-                else if (velocity.X < 0f)
+                    else if (velocity.X < 0f)
-                {
+                    {
-                    direction = -1;
+                        direction = -1;
-                }
+                    }
 
-                mount.Delegations.MinecartBumperSound(position, width, height);
+                    mount.Delegations.MinecartBumperSound(position, width, height);
-            }
+                }
 
-            velocity /= num100;
+                velocity /= num100;
-            if (bitsByte[3] && whoAmI == Main.myPlayer)
+                if (bitsByte[3] && whoAmI == Main.myPlayer)
-            {
+                {
-                flag27 = true;
+                    flag27 = true;
-            }
+                }
 
-            if (bitsByte[2])
+                if (bitsByte[2])
-            {
+                {
-                cartRampTime = (int)(Math.Abs(velocity.X) / mount.RunSpeed * 20f);
+                    cartRampTime = (int)(Math.Abs(velocity.X) / mount.RunSpeed * 20f);
-            }
+                }
 
-            if (bitsByte[4])
+                if (bitsByte[4])
-            {
+                {
-                trackBoost -= 4f;
+                    trackBoost -= 4f;
-            }
+                }
 
-            if (bitsByte[5])
+                if (bitsByte[5])
-            {
+                {
-                trackBoost += 4f;
+                    trackBoost += 4f;
+                }
             }
-        }
 
-        bool flag28 = whoAmI == Main.myPlayer && !mount.Active;
+            bool flag28 = whoAmI == Main.myPlayer && !mount.Active;
-        Vector2 vector5 = position;
+            Vector2 vector5 = position;
-        if (vortexDebuff)
+            if (vortexDebuff)
-        {
+            {
-            velocity.Y = velocity.Y * 0.8f + (float)Math.Cos(base.Center.X % 120f / 120f * ((float)Math.PI * 2f)) * 5f * 0.2f;
+                velocity.Y = velocity.Y * 0.8f + (float)Math.Cos(base.Center.X % 120f / 120f * ((float)Math.PI * 2f)) * 5f * 0.2f;
-        }
+            }
 
-        if (tongued)
+            if (tongued)
-        {
+            {
-            position += velocity;
+                position += velocity;
-            flag28 = false;
+                flag28 = false;
-        }
+            }
-        else if (honeyWet && !ignoreWater)
+            else if (honeyWet && !ignoreWater)
-        {
+            {
-            HoneyCollision(fallThrough, ignorePlats);
+                HoneyCollision(fallThrough, ignorePlats);
-        }
+            }
-        else if (wet && !merman && !ignoreWater && !trident)
+            else if (wet && !merman && !ignoreWater && !trident)
-        {
+            {
-            WaterCollision(fallThrough, ignorePlats);
+                WaterCollision(fallThrough, ignorePlats);
-        }
+            }
-        else
+            else
-        {
+            {
-            DryCollision(fallThrough, ignorePlats);
+                DryCollision(fallThrough, ignorePlats);
-            if (mount.Active && mount.IsConsideredASlimeMount && velocity.Y != 0f && !SlimeDontHyperJump)
+                if (mount.Active && mount.IsConsideredASlimeMount && velocity.Y != 0f && !SlimeDontHyperJump)
-            {
+                {
-                Vector2 vector6 = velocity;
+                    Vector2 vector6 = velocity;
-                velocity.X = 0f;
+                    velocity.X = 0f;
-                DryCollision(fallThrough, ignorePlats);
+                    DryCollision(fallThrough, ignorePlats);
-                velocity.X = vector6.X;
+                    velocity.X = vector6.X;
-            }
+                }
 
-            if (mount.Active && mount.Type == 43 && velocity.Y != 0f)
+                if (mount.Active && mount.Type == 43 && velocity.Y != 0f)
-            {
+                {
-                Vector2 vector7 = velocity;
+                    Vector2 vector7 = velocity;
-                velocity.X = 0f;
+                    velocity.X = 0f;
-                DryCollision(fallThrough, ignorePlats);
+                    DryCollision(fallThrough, ignorePlats);
-                velocity.X = vector7.X;
+                    velocity.X = vector7.X;
+                }
             }
-        }
 
-        UpdateTouchingTiles();
+            UpdateTouchingTiles();
-        TryBouncingBlocks(falling);
+            TryBouncingBlocks(falling);
-        TryLandingOnDetonator();
+            TryLandingOnDetonator();
-        if (!tongued)
+            if (!tongued)
-        {
+            {
-            SlopingCollision(fallThrough, ignorePlats);
+                SlopingCollision(fallThrough, ignorePlats);
-            if (!isLockedToATile)
+                if (!isLockedToATile)
-            {
+                {
-                Collision.StepConveyorBelt(this, gravDir);
+                    Collision.StepConveyorBelt(this, gravDir);
+                }
             }
-        }
 
-        if (flag28 && velocity.Y == 0f)
+            if (flag28 && velocity.Y == 0f)
-        {
+            {
-            AchievementsHelper.HandleRunning(Math.Abs(position.X - vector5.X));
+                AchievementsHelper.HandleRunning(Math.Abs(position.X - vector5.X));
-        }
+            }
 
-        if (flag27)
+            if (flag27)
-        {
+            {
-            NetMessage.SendData(13, -1, -1, null, whoAmI);
+                NetMessage.SendData(13, -1, -1, null, whoAmI);
-            Minecart.HitTrackSwitch(new Vector2(position.X, position.Y), width, height);
+                Minecart.HitTrackSwitch(new Vector2(position.X, position.Y), width, height);
-        }
+            }
 
-        if (vector4.X != velocity.X)
+            if (vector4.X != velocity.X)
-        {
+            {
-            if (vector4.X < 0f)
+                if (vector4.X < 0f)
-            {
+                {
-                slideDir = -1;
+                    slideDir = -1;
-            }
+                }
-            else if (vector4.X > 0f)
+                else if (vector4.X > 0f)
-            {
+                {
-                slideDir = 1;
+                    slideDir = 1;
+                }
             }
-        }
 
-        if (gravDir == 1f && Collision.up)
+            if (gravDir == 1f && Collision.up)
-        {
+            {
-            velocity.Y = 0.01f;
+                velocity.Y = 0.01f;
-            if (!merman)
+                if (!merman)
-            {
+                {
-                jump = 0;
+                    jump = 0;
+                }
             }
-        }
-        else if (gravDir == -1f && Collision.down)
+            else if (gravDir == -1f && Collision.down)
-        {
+            {
-            velocity.Y = -0.01f;
+                velocity.Y = -0.01f;
-            if (!merman)
+                if (!merman)
-            {
+                {
-                jump = 0;
+                    jump = 0;
+                }
             }
-        }
 
-        if (velocity.Y == 0f && grappling[0] == -1)
+            if (velocity.Y == 0f && grappling[0] == -1)
-        {
+            {
-            FloorVisuals(falling);
+                FloorVisuals(falling);
-        }
+            }
 
-        if (whoAmI == Main.myPlayer)
+            if (whoAmI == Main.myPlayer)
-        {
+            {
-            Collision.SwitchTiles(position, width, height, oldPosition, 1);
+                Collision.SwitchTiles(position, width, height, oldPosition, 1);
-        }
+            }
 
-        PressurePlateHelper.UpdatePlayerPosition(this);
+            PressurePlateHelper.UpdatePlayerPosition(this);
-        BordersMovement();
+            BordersMovement();
+        }
         numMinions = 0;
         slotsMinions = 0f;
         if (mount.Type != 8)
