--- src/Terraria/Terraria/Player.cs
+++ src/TerraAngel/Terraria/Player.cs
@@ -533,6 +_,7 @@
         }
     }
 
+    public bool TerraAngelUser = false;
     public int emoteTime;
     public CreativeUnlocksTracker creativeTracker;
     private static byte[] ENCRYPTION_KEY = new UnicodeEncoding().GetBytes("h3y_gUyZ");
@@ -17796,7 +_,7 @@
                 }
 
                 int num5 = tile.frameY / 22;
-                if (num5 < 24)
+                if (num5 < 22 && num5 >= 0)
                 {
                     nearbyTorch[num5] = true;
                     if (num5 == 17 && (tile.liquid == 0 || tile.liquidType() != 0))
@@ -19539,7 +_,7 @@
 
         int num = (int)(position.X + (float)(width / 2)) / 16;
         int num2 = (int)(position.Y - 8f) / 16;
-        if (!WorldGen.IsRope(num, num2))
+        if (!WorldGen.InWorld(num, num2) || Main.tile[num, num2] == null || !Main.tile[num, num2].active() || !Main.tileRope[Main.tile[num, num2].type])
         {
             return;
         }
@@ -23723,7 +_,7 @@
 
     public static int GetFloorTileType(int x, int y)
     {
-        return ((int?)GetFloorTile(x, y)?.type) ?? (-1);
+        return ((int)GetFloorTile(x, y).type);// ?? (-1);
     }
 
     private void MakeFloorDust(bool Falling, int type, int paintColor)
@@ -28330,6 +_,9 @@
             CheckCrackedBrickBreak();
         }
 
+        
+        if ((!CringeManager.GetCringe<NoClipCringe>().Enabled && i == Main.myPlayer) || (i != Main.myPlayer && Main.tile.IsTileInLoadedSection(((int)(position.X / 16f)), ((int)(position.Y / 16f)))))
+        {
         if (!shimmering)
         {
             SlopeDownMovement();
@@ -28560,6 +_,7 @@
         {
             ItemCheck_ManageRightClickFeatures();
         }
+        }
 
         ItemCheckWrapped(i);
         PlayerFrame();
@@ -28584,6 +_,7 @@
         UpdateAdvancedShadows();
     }
 
+
     private void AdjustRemainingPotionSickness()
     {
         if (whoAmI == Main.myPlayer)
@@ -42823,7 +_,7 @@
         return num;
     }
 
-    private void Fishing_GetBait(out int baitPower, out int baitType)
+    public void Fishing_GetBait(out int baitPower, out int baitType)
     {
         baitPower = 0;
         baitType = 0;
