--- src/Terraria/Terraria/Program.cs
+++ src/TerraAngel/Terraria/Program.cs
@@ -6,19 +_,18 @@
 using System.Runtime.CompilerServices;
 using System.Text;
 using System.Threading;
-using System.Windows.Forms;
 using ReLogic.IO;
 using ReLogic.OS;
+using TerraAngel;
 using Terraria.Initializers;
 using Terraria.Localization;
-using Terraria.Social;
 using Terraria.Utilities;
 
 namespace Terraria;
 public static class Program
 {
-    public static bool IsXna = true;
+    public static bool IsXna = false;
-    public static bool IsFna = false;
+    public static bool IsFna = true;
     public static bool IsMono = Type.GetType("Mono.Runtime") != null;
     public const bool IsDebug = false;
     public static Dictionary<string, string> LaunchParameters = new Dictionary<string, string>();
@@ -187,6 +_,7 @@
 
         LaunchParameters = Utils.ParseArguements(args);
         SavePath = (LaunchParameters.ContainsKey("-savedirectory") ? LaunchParameters["-savedirectory"] : Platform.Get<IPathService>().GetStoragePath("Terraria"));
+        ClientLoader.LoadClient();
         ThreadPool.SetMinThreads(8, 8);
         InitializeConsoleOutput();
         SetupLogging();
@@ -206,10 +_,12 @@
         }
 
         using Main main = new Main();
+#if !DEBUG
         try
         {
+#endif
             Lang.InitializeLegacyLocalization();
-            SocialAPI.Initialize();
+            //SocialAPI.Initialize();
             LaunchInitializer.LoadParameters(main);
             Main.OnEnginePreload += StartForceLoad;
             if (Main.dedServ)
@@ -218,11 +_,13 @@
             }
 
             main.Run();
+#if !DEBUG
         }
         catch (Exception e)
         {
             DisplayException(e);
         }
+#endif
     }
 
     private static void TrySettingFNAToOpenGL(string[] args)
@@ -243,7 +_,7 @@
         }
     }
 
-    private static void DisplayException(Exception e)
+    public static void DisplayException(Exception e)
     {
         try
         {
@@ -271,7 +_,7 @@
                 Console.WriteLine(Language.GetTextValue("Error.ServerCrash"), DateTime.Now, text);
             }
 
-            MessageBox.Show(text, "Terraria: Error");
+            SDL2.SDL.SDL_ShowSimpleMessageBox(SDL2.SDL.SDL_MessageBoxFlags.SDL_MESSAGEBOX_ERROR, "Terraria: Error", text, Main.instance.Window.Handle);
         }
         catch
         {
