--- src/Terraria/Terraria/Terraria.csproj
+++ src/TerraAngel/Terraria/Terraria.csproj
@@ -3,40 +_,132 @@
   <PropertyGroup>
     <OutputType>WinExe</OutputType>
     <Nullable>enable</Nullable>
+    <Deterministic>false</Deterministic>
     <Version>1.4.3.6</Version>
     <Company>Re-Logic</Company>
     <Company>Copyright © 2022 Re-Logic</Company>
     <RootNamespace>Terraria</RootNamespace>
+    <TargetFramework>net6.0</TargetFramework>
+    <Platforms>AnyCPU</Platforms>
   </PropertyGroup>
+  <PropertyGroup>
+    <ApplicationIcon>app.ico</ApplicationIcon>
+    <Win32Resource />
+  </PropertyGroup>
+
+  <!-- Libraries part of Terraria -->
   <ItemGroup>
-    <Reference Include="CsvHelper" />
-    <Reference Include="Ionic.Zip.CF" />
-    <Reference Include="Microsoft.Xna.Framework" />
-    <Reference Include="Microsoft.Xna.Framework.Game" />
-    <Reference Include="Microsoft.Xna.Framework.Graphics" />
-    <Reference Include="Microsoft.Xna.Framework.Xact" />
-    <Reference Include="MP3Sharp" />
-    <Reference Include="Newtonsoft.Json" />
-    <Reference Include="NVorbis" />
-    <Reference Include="RailSDK.Net" />
+    <Reference Include="Libraries/CsvHelper/**" />
+    <Reference Include="Libraries/DotNetZip/**" />
+    <Reference Include="Libraries/MP3Sharp/**" />
+    <Reference Include="Libraries/NVorbis/**" />
     <ProjectReference Include="../ReLogic/ReLogic.csproj" />
-    <EmbeddedResource Include="../ReLogic/bin/$(Configuration)/$(TargetFramework)/ReLogic.dll" />
-    <Reference Include="Steamworks.NET" />
-    <Reference Include="SteelSeriesEngineWrapper" />
-    <Reference Include="System" />
-    <Reference Include="System.Core" />
-    <Reference Include="System.Drawing" />
-    <Reference Include="System.Runtime.Serialization" />
-    <Reference Include="System.Windows.Forms" />
-    <Reference Include="WindowsBase" />
-  </ItemGroup>
+  </ItemGroup>
+
+  
+
+  <!-- New libraries added -->
+  <ItemGroup>
+    <Reference Include="FNA">
+      <HintPath>LibNew\FNA\FNA.dll</HintPath>
+    </Reference>
+    <Reference Include="ImGui.NET">
+      <HintPath>LibNew\ImGui\ImGui.NET.dll</HintPath>
+    </Reference>
+    <Reference Include="Mono.Cecil">
+      <HintPath>LibNew\MonoMod\Mono.Cecil.dll</HintPath>
+    </Reference>
+    <Reference Include="MonoMod.RuntimeDetour">
+      <HintPath>LibNew\MonoMod\MonoMod.RuntimeDetour.dll</HintPath>
+    </Reference>
+    <Reference Include="MonoMod.Utils">
+      <HintPath>LibNew\MonoMod\MonoMod.Utils.dll</HintPath>
+    </Reference>
+    <Reference Include="LibNew/x64/Steamworks/Steamworks.NET.dll" />
+    <ProjectReference Include="../TerraAngelPluginAPI/TerraAngelPluginAPI.csproj" />
+  </ItemGroup>
+
   <ItemGroup>
     <EmbeddedResource Include="GameContent/Creative/Content/*" />
     <EmbeddedResource Include="GameContent/Metadata/MaterialData/*" />
     <EmbeddedResource Include="GameContent/WorldBuilding/*" />
     <EmbeddedResource Include="IO/Data/*" />
-    <EmbeddedResource Include="Libraries/**" />
     <EmbeddedResource Include="Localization/Content/**" />
-    <EmbeddedResource Include="Microsoft/**" />
+  </ItemGroup>
+
+  <Target Name="EditBin" AfterTargets="Build">
+    <PropertyGroup>
+      <EditBinOpts>/largeaddressaware</EditBinOpts>
+      <DumpBin>/headers</DumpBin>
+    </PropertyGroup>
+
+    <Message Text="Applying editbin $(EditBinOpts)" Importance="high" />
+    <Exec Command="..\largeaddresstool\editbin $(EditBinOpts) &quot;$(TargetPath)&quot;" />
+    <Message Text="Checking headers with dumpbin." Importance="high" />
+    <Exec Command="..\largeaddresstool\dumpbin $(DumpBin) &quot;$(TargetPath)&quot; | findstr &quot;large&quot;" />
+  </Target>
+
+
+  <ItemGroup>
+    <PackageReference Include="DiscordRichPresence" Version="1.0.175" />
+    <PackageReference Include="Microsoft.CodeAnalysis" Version="4.2.0" />
+    <PackageReference Include="Microsoft.CodeAnalysis.CSharp.Features" Version="4.2.0" />
+    <PackageReference Include="Microsoft.CodeAnalysis.CSharp.Scripting" Version="4.2.0" />
+    <PackageReference Include="Microsoft.CodeAnalysis.Features" Version="4.2.0" />
+    <PackageReference Include="Microsoft.CodeAnalysis.Workspaces.Common" Version="4.2.0" />
+    <PackageReference Include="System.Drawing.Common" Version="6.0.0" />
+    <PackageReference Include="Newtonsoft.Json" Version="13.0.1" />
+  </ItemGroup>
+
+
+  <ItemGroup>
+    <None Update="Assets\IconFont.ttf">
+      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
+    </None>
+    <None Update="Assets\MonoFont.ttf">
+      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
+    </None>
+    <None Update="Assets\RussianGlyphFont.ttf">
+      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
+    </None>
+    <None Update="Assets\TerrariaFont.ttf">
+      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
+    </None>
+    <None Update="LibNew\x64\FNA\FAudio.dll">
+      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
+    </None>
+    <None Update="LibNew\x64\FNA\FNA3D.dll">
+      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
+    </None>
+    <None Update="LibNew\x64\FNA\libtheorafile.dll">
+      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
+    </None>
+    <None Update="LibNew\x64\FNA\SDL2.dll">
+      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
+    </None>
+    <None Update="LibNew\x64\ImGui\cimgui.dll">
+      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
+    </None>
+    <None Update="LibNew\x64\Steamworks\steam_api.dll">
+      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
+    </None>
+    <None Update="LibNew\x86\FNA\FAudio.dll">
+      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
+    </None>
+    <None Update="LibNew\x86\FNA\FNA3D.dll">
+      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
+    </None>
+    <None Update="LibNew\x86\FNA\libtheorafile.dll">
+      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
+    </None>
+    <None Update="LibNew\x86\FNA\SDL2.dll">
+      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
+    </None>
+    <None Update="LibNew\x86\ImGui\cimgui.dll">
+      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
+    </None>
+    <None Update="LibNew\x86\Steamworks\steam_api.dll">
+      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
+    </None>
   </ItemGroup>
 </Project>
