--- src/Terraria/Terraria/Terraria.csproj
+++ src/TerraAngel/Terraria/Terraria.csproj
@@ -7,29 +_,34 @@
     <Company>Re-Logic</Company>
     <Company>Copyright © 2022 Re-Logic</Company>
     <RootNamespace>Terraria</RootNamespace>
+    <TargetFramework>net6.0</TargetFramework>
+    <Platforms>x86;x64</Platforms>
   </PropertyGroup>
+  <PropertyGroup>
+    <ApplicationIcon>app.ico</ApplicationIcon>
+    <Win32Resource />
+  </PropertyGroup>
+
+  <!-- Libraries part of Terraria -->
   <ItemGroup>
-    <Reference Include="CsvHelper" />
-    <Reference Include="Ionic.Zip.CF" />
-    <Reference Include="Microsoft.Xna.Framework" />
-    <Reference Include="Microsoft.Xna.Framework.Game" />
-    <Reference Include="Microsoft.Xna.Framework.Graphics" />
-    <Reference Include="Microsoft.Xna.Framework.Xact" />
-    <Reference Include="MP3Sharp" />
-    <Reference Include="Newtonsoft.Json" />
-    <Reference Include="NVorbis" />
-    <Reference Include="RailSDK.Net" />
+    <Reference Include="Libraries/CsvHelper/**" />
+    <Reference Include="Libraries/DotNetZip/**" />
+    <Reference Include="Libraries/MP3Sharp/**" />
+    <Reference Include="Libraries/NVorbis/**" />
     <ProjectReference Include="../ReLogic/ReLogic.csproj" />
-    <EmbeddedResource Include="../ReLogic/bin/$(Configuration)/$(TargetFramework)/ReLogic.dll" />
-    <Reference Include="Steamworks.NET" />
-    <Reference Include="SteelSeriesEngineWrapper" />
-    <Reference Include="System" />
-    <Reference Include="System.Core" />
-    <Reference Include="System.Drawing" />
-    <Reference Include="System.Runtime.Serialization" />
-    <Reference Include="System.Windows.Forms" />
-    <Reference Include="WindowsBase" />
-  </ItemGroup>
+    <EmbeddedResource Include="../ReLogic/bin/$(Platform)/$(Configuration)/$(TargetFramework)/ReLogic.dll" />
+  </ItemGroup>
+
+  <!-- New libraries added -->
+  <ItemGroup>
+    <Reference Include="LibNew/$(Platform)/FNA/FNA.dll" />
+    <Reference Include="LibNew/$(Platform)/Steamworks.NET.dll" />
+    <Reference Include="LibNew/$(Platform)/ImGui/ImGui.NET.dll" />
+    <Reference Include="LibNew/AnyCpu/MonoMod/Mono.Cecil.dll" />
+    <Reference Include="LibNew/AnyCpu/MonoMod/MonoMod.Utils.dll" />
+    <Reference Include="LibNew/AnyCpu/MonoMod/MonoMod.RuntimeDetour.dll" />
+  </ItemGroup>
+
   <ItemGroup>
     <EmbeddedResource Include="GameContent/Creative/Content/*" />
     <EmbeddedResource Include="GameContent/Metadata/MaterialData/*" />
@@ -38,5 +_,32 @@
     <EmbeddedResource Include="Libraries/**" />
     <EmbeddedResource Include="Localization/Content/**" />
     <EmbeddedResource Include="Microsoft/**" />
+    <EmbeddedResource Include="LibNew/$(Platform)/FNA/FNA.dll" />
+    <EmbeddedResource Include="LibNew/$(Platform)/Steamworks.NET.dll" />
   </ItemGroup>
+	
+  <!-- Copy some steam stuff -->
+  <Target Name="CopySteam" AfterTargets="AfterBuild">
+    <Copy SourceFiles="LibNew/$(Platform)/steam_api.dll" DestinationFolder="$(OutDir)" />
+  </Target>
+  
+
+  <!-- Copy some FNA stuff -->
+  <Target Name="CopyFNA" AfterTargets="AfterBuild">
+    <Copy SourceFiles="LibNew/$(Platform)/FNA/FAudio.dll" DestinationFolder="$(OutDir)" />
+    <Copy SourceFiles="LibNew/$(Platform)/FNA/FNA3D.dll" DestinationFolder="$(OutDir)" />
+    <Copy SourceFiles="LibNew/$(Platform)/FNA/SDL2.dll" DestinationFolder="$(OutDir)" />
+    <Copy SourceFiles="LibNew/$(Platform)/FNA/libtheorafile.dll" DestinationFolder="$(OutDir)" />
+  </Target>
+
+  <!-- Copy some ImGui stuff -->
+  <Target Name="CopyImGui" AfterTargets="AfterBuild">
+    <Copy SourceFiles="LibNew/$(Platform)/ImGui/cimgui.dll" DestinationFolder="$(OutDir)" />
+  </Target>
+
+  <ItemGroup>
+    <PackageReference Include="System.Drawing.Common" Version="6.0.0" />
+    <PackageReference Include="Newtonsoft.Json" Version="13.0.1" />
+  </ItemGroup>
+
 </Project>
