--- src/Terraria/Terraria/Utils.cs
+++ src/TerraAngel/Terraria/Utils.cs
@@ -1986,7 +_,7 @@
 
     public static string ReadEmbeddedResource(string path)
     {
-        using Stream stream = Assembly.GetExecutingAssembly().GetManifestResourceStream(path);
+        using Stream stream = Assembly.GetAssembly(typeof(Main)).GetManifestResourceStream(path);
         using StreamReader streamReader = new StreamReader(stream);
         return streamReader.ReadToEnd();
     }
