--- src/Terraria/Terraria/Utils.cs
+++ src/TerraAngel/Terraria/Utils.cs
@@ -53,6 +_,11 @@
         return value1 + (value2 - value1) * amount;
     }
 
+    public static float Lerp(float value1, float value2, float amount)
+    {
+        return value1 + (value2 - value1) * amount;
+    }
+
     public static Vector2 Round(Vector2 input)
     {
         return new Vector2((float)Math.Round(input.X), (float)Math.Round(input.Y));
@@ -1997,7 +_,7 @@
 
     public static string ReadEmbeddedResource(string path)
     {
-        using Stream stream = Assembly.GetExecutingAssembly().GetManifestResourceStream(path);
+        using Stream stream = Assembly.GetAssembly(typeof(Main)).GetManifestResourceStream(path);
         using StreamReader streamReader = new StreamReader(stream);
         return streamReader.ReadToEnd();
     }
