--- src/Terraria/Terraria/WorldGen.cs
+++ src/TerraAngel/Terraria/WorldGen.cs
@@ -3552,15 +_,9 @@
         LiquidBuffer.numLiquidBuffer = 0;
         if (Main.netMode == 1 || lastMaxTilesX > Main.maxTilesX || lastMaxTilesY > Main.maxTilesY)
         {
-            for (int j = 0; j < lastMaxTilesX; j++)
-            {
-                float num = (float)j / (float)lastMaxTilesX;
-                Main.statusText = Lang.gen[46].Value + " " + (int)(num * 100f + 1f) + "%";
-                for (int k = 0; k < lastMaxTilesY; k++)
-                {
-                    Main.tile[j, k] = null;
-                }
-            }
+            Main.statusText = Lang.gen[46].Value + "0%";
+            Main.tile.ResetHeap();
+            Main.statusText = Lang.gen[46].Value + "100%";
         }
 
         lastMaxTilesX = Main.maxTilesX;
@@ -3572,22 +_,9 @@
 
         if (Main.netMode != 1)
         {
-            for (int l = 0; l < Main.maxTilesX; l++)
-            {
-                float num2 = (float)l / (float)Main.maxTilesX;
-                Main.statusText = Lang.gen[47].Value + " " + (int)(num2 * 100f + 1f) + "%";
-                for (int m = 0; m < Main.maxTilesY; m++)
-                {
-                    if (Main.tile[l, m] == null)
-                    {
-                        Main.tile[l, m] = new Tile();
-                    }
-                    else
-                    {
-                        Main.tile[l, m].ClearEverything();
-                    }
-                }
-            }
+            Main.statusText = Lang.gen[46].Value + "0%";
+            Main.tile.ResetHeap();
+            Main.statusText = Lang.gen[46].Value + "100%";
         }
 
         for (int n = 0; n < Main.countsAsHostForGameplay.Length; n++)
@@ -66054,6 +_,7 @@
 
     public static bool SolidOrSlopedTile(int x, int y)
     {
+        if (!Main.tile.IsTileInLoadedSection(x, y)) return true;
         return SolidOrSlopedTile(Main.tile[x, y]);
     }
 
@@ -66084,13 +_,9 @@
 
     public static bool SolidTile(int i, int j, bool noDoors = false)
     {
+        if (!Main.tile.IsTileInLoadedSection(i, j)) return true;
         try
         {
-            if (Main.tile[i, j] == null)
-            {
-                return true;
-            }
-
             if (Main.tile[i, j].active() && Main.tileSolid[Main.tile[i, j].type] && !Main.tileSolidTop[Main.tile[i, j].type] && !Main.tile[i, j].halfBrick() && Main.tile[i, j].slope() == 0 && !Main.tile[i, j].inActive())
             {
                 if (noDoors && Main.tile[i, j].type == 10)
@@ -83128,6 +_,7 @@
         {
             for (int j = 0; j < height; j++)
             {
+                if (!WorldGen.InWorld(x + i, y + j)) continue;
                 Tile tile = Main.tile[x + i, y + j];
                 if (tile == null)
                 {
