--- src/Terraria/Terraria/GameContent/UI/ResourceSets/ResourceDrawSettings.cs
+++ src/TerraAngel/Terraria/GameContent/UI/ResourceSets/ResourceDrawSettings.cs
@@ -39,7 +_,7 @@
                 isHovered = true;
             }
 
-            spriteBatch.Draw(texture.Value, position, rectangle, Color.White, 0f, origin, drawScale, SpriteEffects.None, 0f);
+            spriteBatch.Draw(texture.Value, position + new Vector2(0, 8f), rectangle, Color.White, 0f, origin, drawScale, SpriteEffects.None, 0f);
             topLeftAnchor += OffsetPerDraw + rectangle.Size() * OffsetPerDrawByTexturePercentile;
         }
     }
