--- src/Terraria/Terraria/GameContent/ObjectInteractions/AHoverInteractionChecker.cs
+++ src/TerraAngel/Terraria/GameContent/ObjectInteractions/AHoverInteractionChecker.cs
@@ -20,9 +_,8 @@
             return HoverStatus.NotSelectable;
         }
 
-        Matrix matrix = Matrix.Invert(Main.GameViewMatrix.ZoomMatrix);
         Vector2 vector = Main.ReverseGravitySupport(Main.MouseScreen);
-        Vector2.Transform(Main.screenPosition, matrix);
+        Vector2.Transform(Main.screenPosition, Main.GameViewMatrix.InverseZoomMatrix);
         Vector2 v = vector + Main.screenPosition;
         bool flag = Hitbox.Contains(v.ToPoint());
         bool flag2 = flag;
