--- src/Terraria/Terraria/ID/MessageID.cs
+++ src/TerraAngel/Terraria/ID/MessageID.cs
@@ -21,15 +_,15 @@
         public const byte PlayerActive = 14;
         [Old("Deprecated.")]
         public const byte Unknown15 = 15;
-        public const byte PlayerLifeMana = 16;
+        public const byte PlayerLife = 16;
         public const byte TileManipulation = 17;
         public const byte SetTime = 18;
         public const byte ToggleDoorState = 19;
-        public const byte Unknown20 = 20;
+        public const byte SendTileSquare = 20;
         public const byte SyncItem = 21;
         public const byte ItemOwner = 22;
         public const byte SyncNPC = 23;
-        public const byte UnusedMeleeStrike = 24;
+        public const byte StrikeNPCWithHeldItem = 24;
         [Old("Deprecated. Use NetTextModule instead.")]
         public const byte Unused25 = 25;
         [Old("Deprecated.")]
@@ -48,36 +_,36 @@
         public const byte SendPassword = 38;
         public const byte ReleaseItemOwnership = 39;
         public const byte SyncTalkNPC = 40;
-        public const byte Unknown41 = 41;
+        public const byte PlayerItemAnimation = 41;
-        public const byte Unknown42 = 42;
+        public const byte PlayerMana = 42;
-        public const byte Unknown43 = 43;
+        public const byte ManaEffect = 43;
         [Old("Deprecated.")]
         public const byte Unknown44 = 44;
-        public const byte Unknown45 = 45;
+        public const byte PlayerTeam = 45;
-        public const byte Unknown46 = 46;
+        public const byte RequestSign = 46;
-        public const byte Unknown47 = 47;
+        public const byte UpdateSign = 47;
         [Old("Deprecated. Use NetLiquidModule instead.")]
         public const byte LiquidUpdate = 48;
-        public const byte Unknown49 = 49;
+        public const byte CompleteConnectionAndSpawn = 49;
-        public const byte Unknown50 = 50;
+        public const byte UpdatePlayerBuff = 50;
         public const byte MiscDataSync = 51;
-        public const byte Unknown52 = 52;
+        public const byte UnlockTile = 52;
-        public const byte Unknown53 = 53;
+        public const byte AddNPCBuff = 53;
-        public const byte Unknown54 = 54;
+        public const byte UpdateNPCBuff = 54;
         public const byte AddPlayerBuff = 55;
         public const byte UniqueTownNPCInfoSyncRequest = 56;
-        public const byte Unknown57 = 57;
+        public const byte UpdateWorldEvil = 57;
         public const byte InstrumentSound = 58;
         public const byte HitSwitch = 59;
-        public const byte Unknown60 = 60;
+        public const byte UpdateNPCHome = 60;
-        public const byte Unknown61 = 61;
+        public const byte SpawnBoss = 61;
-        public const byte Unknown62 = 62;
+        public const byte PlayerDodge = 62;
-        public const byte Unknown63 = 63;
+        public const byte PaintTile = 63;
-        public const byte Unknown64 = 64;
+        public const byte PaintWall = 64;
         public const byte TeleportEntity = 65;
-        public const byte Unknown66 = 66;
+        public const byte PlayerNPCTeleport = 66;
         public const byte Unknown67 = 67;
-        public const byte Unknown68 = 68;
+        public const byte ClientUUID = 68;
         public const byte ChestName = 69;
         public const byte BugCatching = 70;
         public const byte BugReleasing = 71;
