--- src/Terraria/Terraria/Net/Sockets/TcpSocket.cs
+++ src/TerraAngel/Terraria/Net/Sockets/TcpSocket.cs
@@ -17,6 +_,7 @@
     private SocketConnectionAccepted _listenerCallback;
     private RemoteAddress _remoteAddress;
     private bool _isListening;
+    private NetworkStream networkStream;
     public int MessagesInQueue => _messagesInQueue;
     public TcpSocket()
     {
@@ -52,20 +_,27 @@
         TcpAddress tcpAddress = (TcpAddress)address;
         _connection.Connect(tcpAddress.Address, tcpAddress.Port);
         _remoteAddress = address;
+        networkStream = _connection.GetStream();
     }
 
     private void ReadCallback(IAsyncResult result)
     {
         Tuple<SocketReceiveCallback, object> tuple = (Tuple<SocketReceiveCallback, object>)result.AsyncState;
-        tuple.Item1(tuple.Item2, _connection.GetStream().EndRead(result));
+        int messageLength = networkStream.EndRead(result);
+        StatsWindow.CountReadMessage(messageLength);
+        tuple.Item1(tuple.Item2, messageLength);
     }
 
     private void SendCallback(IAsyncResult result)
     {
         Tuple<SocketSendCallback, object> tuple;
+        int messageLength = 0;
         if (Platform.IsWindows)
         {
+            object[] data = (object[])result.AsyncState!;
-            tuple = (Tuple<SocketSendCallback, object>)result.AsyncState;
+            tuple = (Tuple<SocketSendCallback, object>)data[0];
+            messageLength = (int)data[1];
+            StatsWindow.CountSentMessage(messageLength);
         }
         else
         {
@@ -76,13 +_,14 @@
 
         try
         {
-            _connection.GetStream().EndWrite(result);
+            networkStream.EndWrite(result);
             tuple.Item1(tuple.Item2);
         }
         catch (Exception)
         {
             ((ISocket)this).Close();
         }
+
     }
 
     void ISocket.SendQueuedPackets()
@@ -94,17 +_,17 @@
         if (!Platform.IsWindows)
         {
             byte[] array = LegacyNetBufferPool.RequestBuffer(data, offset, size);
-            _connection.GetStream().BeginWrite(array, 0, size, SendCallback, new object[2] { new Tuple<SocketSendCallback, object>(callback, state), array });
+            networkStream.BeginWrite(array, 0, size, SendCallback, new object[] { new Tuple<SocketSendCallback, object>(callback, state), array });
         }
         else
         {
-            _connection.GetStream().BeginWrite(data, 0, size, SendCallback, new Tuple<SocketSendCallback, object>(callback, state));
+            networkStream.BeginWrite(data, 0, size, SendCallback, new object[] { new Tuple<SocketSendCallback, object>(callback, state), size });
         }
     }
 
     void ISocket.AsyncReceive(byte[] data, int offset, int size, SocketReceiveCallback callback, object state)
     {
-        _connection.GetStream().BeginRead(data, offset, size, ReadCallback, new Tuple<SocketReceiveCallback, object>(callback, state));
+        networkStream.BeginRead(data, offset, size, ReadCallback, new Tuple<SocketReceiveCallback, object>(callback, state));
     }
 
     bool ISocket.IsDataAvailable()
@@ -114,7 +_,7 @@
             return false;
         }
 
-        return _connection.GetStream().DataAvailable;
+        return networkStream.DataAvailable;
     }
 
     RemoteAddress ISocket.GetRemoteAddress()
