--- src/Terraria/Terraria/GameInput/PlayerInput.cs
+++ src/TerraAngel/Terraria/GameInput/PlayerInput.cs
@@ -595,7 +_,9 @@
         return _buttonsLocked.Contains(keyName);
     }
 
+    public static void UpdateInput() => TerraAngel.Hooks.Hooks.MiscHooks.UpdateInputHook(_UpdateInput);
+
-    public static void UpdateInput()
+    private static void _UpdateInput()
     {
         SettingsForUI.UpdateCounters();
         Triggers.Reset();
@@ -1056,8 +_,6 @@
                     MouseY = num10 + num5;
                     flag6 = true;
                 }
-
-                flag = true;
             }
 
             if (CurrentInputMode == InputMode.XBoxGamepad)
@@ -1145,7 +_,9 @@
         return flag;
     }
 
+    private static void MouseInput() => TerraAngel.Hooks.Hooks.MiscHooks.MouseInputHook(_MouseInput);
+
-    private static void MouseInput()
+    private static void _MouseInput()
     {
         bool flag = false;
         MouseInfoOld = MouseInfo;
@@ -1227,8 +_,6 @@
             {
                 flag2 = true;
             }
-
-            Main.ChromaPainter.PressKey(pressedKeys[i]);
         }
 
         if (Main.blockKey != Keys.None.ToString())
@@ -1429,7 +_,6 @@
             ListenFor(null, InputMode.Keyboard);
             Main.blockKey = newKey;
             Main.blockInput = false;
-            Main.ChromaPainter.CollectBoundKeys();
         }
 
         if (CurrentlyRebinding && _listeningInputMode == InputMode.KeyboardUI)
@@ -1449,7 +_,6 @@
             ListenFor(null, InputMode.KeyboardUI);
             Main.blockKey = newKey;
             Main.blockInput = false;
-            Main.ChromaPainter.CollectBoundKeys();
         }
 
         FixDerpedRebinds();
@@ -1548,10 +_,9 @@
         {
             Main.SmartCursorWanted_GamePad = false;
             Matrix zoomMatrix = Main.GameViewMatrix.ZoomMatrix;
-            Matrix matrix = Matrix.Invert(zoomMatrix);
             Vector2 mouseScreen = Main.MouseScreen;
-            Vector2.Transform(Main.screenPosition, matrix);
+            Vector2.Transform(Main.screenPosition, Main.GameViewMatrix.InverseZoomMatrix);
-            Vector2 vector2 = Vector2.Transform((Vector2.Transform(mouseScreen, matrix) + zero * new Vector2(16f) + Main.screenPosition).ToTileCoordinates().ToWorldCoordinates() - Main.screenPosition, zoomMatrix);
+            Vector2 vector2 = Vector2.Transform((Vector2.Transform(mouseScreen, Main.GameViewMatrix.InverseZoomMatrix) + zero * new Vector2(16f) + Main.screenPosition).ToTileCoordinates().ToWorldCoordinates() - Main.screenPosition, zoomMatrix);
             MouseX = (int)vector2.X;
             MouseY = (int)vector2.Y;
             SettingsForUI.SetCursorMode(CursorMode.Gamepad);
