--- src/Terraria/Terraria/GameInput/PlayerInput.cs
+++ src/TerraAngel/Terraria/GameInput/PlayerInput.cs
@@ -998,8 +_,6 @@
                     MouseY = num10 + num5;
                     flag6 = true;
                 }
-
-                flag = true;
             }
 
             if (CurrentInputMode == InputMode.XBoxGamepad)
@@ -1169,8 +_,6 @@
             {
                 flag2 = true;
             }
-
-            Main.ChromaPainter.PressKey(pressedKeys[i]);
         }
 
         if (Main.blockKey != Keys.None.ToString())
@@ -1371,7 +_,6 @@
             ListenFor(null, InputMode.Keyboard);
             Main.blockKey = newKey;
             Main.blockInput = false;
-            Main.ChromaPainter.CollectBoundKeys();
         }
 
         if (CurrentlyRebinding && _listeningInputMode == InputMode.KeyboardUI)
@@ -1391,7 +_,6 @@
             ListenFor(null, InputMode.KeyboardUI);
             Main.blockKey = newKey;
             Main.blockInput = false;
-            Main.ChromaPainter.CollectBoundKeys();
         }
 
         FixDerpedRebinds();
