--- src/Terraria/Terraria/Social/SocialAPI.cs
+++ src/TerraAngel/Terraria/Social/SocialAPI.cs
@@ -1,9 +_,8 @@
 using System;
+using System.IO;
 using System.Collections.Generic;
 using ReLogic.OS;
 using Terraria.Social.Base;
-using Terraria.Social.Steam;
-using Terraria.Social.WeGame;
 
 namespace Terraria.Social;
 public static class SocialAPI
@@ -17,7 +_,7 @@
     public static Terraria.Social.Base.WorkshopSocialModule Workshop;
     public static ServerJoinRequestsManager JoinRequests;
     public static Terraria.Social.Base.PlatformSocialModule Platform;
-    private static List<ISocialModule> _modules;
+    private static List<ISocialModule> _modules = new List<ISocialModule>();
     public static SocialMode Mode => _mode;
     public static void Initialize(SocialMode? mode = null)
     {
@@ -36,8 +_,12 @@
                 mode = SocialMode.Steam;
             }
         }
-
-        _mode = mode.Value;
+        if (File.Exists("steam_api.dll"))
+        {
+            _mode = SocialMode.Steam;
+        } else { 
+            _mode = mode.Value;
+        }
         _modules = new List<ISocialModule>();
         JoinRequests = new ServerJoinRequestsManager();
         Main.OnTickForInternalCodeOnly += JoinRequests.Update;
@@ -62,7 +_,7 @@
         _modules.Reverse();
         foreach (ISocialModule module in _modules)
         {
-            module.Shutdown();
+        	module.Shutdown();
         }
     }
 
@@ -107,24 +_,10 @@
             Network = LoadModule<Terraria.Social.Steam.NetClientSocialModule>();
         }
 
-        WeGameHelper.WriteDebugString("LoadSteam modules");
     }
 
     private static void LoadWeGame()
     {
-        LoadModule<Terraria.Social.WeGame.CoreSocialModule>();
-        Cloud = LoadModule<Terraria.Social.WeGame.CloudSocialModule>();
-        Friends = LoadModule<Terraria.Social.WeGame.FriendsSocialModule>();
-        Overlay = LoadModule<Terraria.Social.WeGame.OverlaySocialModule>();
-        if (Main.dedServ)
-        {
-            Network = LoadModule<Terraria.Social.WeGame.NetServerSocialModule>();
-        }
-        else
-        {
-            Network = LoadModule<Terraria.Social.WeGame.NetClientSocialModule>();
-        }
 
-        WeGameHelper.WriteDebugString("LoadWeGame modules");
     }
 }
