--- src/Terraria/Terraria/Social/SocialAPI.cs
+++ src/TerraAngel/Terraria/Social/SocialAPI.cs
@@ -1,7 +_,5 @@
 using System.Collections.Generic;
 using Terraria.Social.Base;
-using Terraria.Social.Steam;
-using Terraria.Social.WeGame;
 
 namespace Terraria.Social;
 public static class SocialAPI
@@ -47,11 +_,10 @@
 
     public static void Shutdown()
     {
-        _modules.Reverse();
-        foreach (ISocialModule module in _modules)
-        {
-            module.Shutdown();
-        }
+        //_modules.Reverse();
+        //foreach (ISocialModule module in _modules) {
+        //	module.Shutdown();
+        //}
     }
 
     private static T LoadModule<T>()
@@ -83,16 +_,14 @@
         Workshop = LoadModule<Terraria.Social.Steam.WorkshopSocialModule>();
         Platform = LoadModule<Terraria.Social.Steam.PlatformSocialModule>();
         Network = LoadModule<Terraria.Social.Steam.NetClientSocialModule>();
-        WeGameHelper.WriteDebugString("LoadSteam modules");
     }
 
     private static void LoadWeGame()
     {
-        LoadModule<Terraria.Social.WeGame.CoreSocialModule>();
+        //LoadModule<Terraria.Social.WeGame.CoreSocialModule>();
-        Cloud = LoadModule<Terraria.Social.WeGame.CloudSocialModule>();
+        //Cloud = LoadModule<Terraria.Social.WeGame.CloudSocialModule>();
-        Friends = LoadModule<Terraria.Social.WeGame.FriendsSocialModule>();
+        //Friends = LoadModule<Terraria.Social.WeGame.FriendsSocialModule>();
-        Overlay = LoadModule<Terraria.Social.WeGame.OverlaySocialModule>();
+        //Overlay = LoadModule<Terraria.Social.WeGame.OverlaySocialModule>();
-        Network = LoadModule<Terraria.Social.WeGame.NetClientSocialModule>();
+        //Network = LoadModule<Terraria.Social.WeGame.NetClientSocialModule>();
-        WeGameHelper.WriteDebugString("LoadWeGame modules");
     }
 }
