--- src/Terraria/Terraria/Social/SocialAPI.cs
+++ src/TerraAngel/Terraria/Social/SocialAPI.cs
@@ -2,8 +_,6 @@
 using System.Collections.Generic;
 using ReLogic.OS;
 using Terraria.Social.Base;
-using Terraria.Social.Steam;
-using Terraria.Social.WeGame;
 
 namespace Terraria.Social;
 public static class SocialAPI
@@ -59,11 +_,10 @@
 
     public static void Shutdown()
     {
-        _modules.Reverse();
-        foreach (ISocialModule module in _modules)
-        {
-            module.Shutdown();
-        }
+        //_modules.Reverse();
+        //foreach (ISocialModule module in _modules) {
+        //	module.Shutdown();
+        //}
     }
 
     private static T LoadModule<T>()
@@ -107,24 +_,9 @@
             Network = LoadModule<Terraria.Social.Steam.NetClientSocialModule>();
         }
 
-        WeGameHelper.WriteDebugString("LoadSteam modules");
     }
 
     private static void LoadWeGame()
     {
-        LoadModule<Terraria.Social.WeGame.CoreSocialModule>();
-        Cloud = LoadModule<Terraria.Social.WeGame.CloudSocialModule>();
-        Friends = LoadModule<Terraria.Social.WeGame.FriendsSocialModule>();
-        Overlay = LoadModule<Terraria.Social.WeGame.OverlaySocialModule>();
-        if (Main.dedServ)
-        {
-            Network = LoadModule<Terraria.Social.WeGame.NetServerSocialModule>();
-        }
-        else
-        {
-            Network = LoadModule<Terraria.Social.WeGame.NetClientSocialModule>();
-        }
-
-        WeGameHelper.WriteDebugString("LoadWeGame modules");
     }
 }
