--- src/Terraria/Terraria/Social/SocialAPI.cs
+++ src/TerraAngel/Terraria/Social/SocialAPI.cs
@@ -2,8 +_,6 @@
 using System.Collections.Generic;
 using ReLogic.OS;
 using Terraria.Social.Base;
-using Terraria.Social.Steam;
-using Terraria.Social.WeGame;
 
 namespace Terraria.Social;
 public static class SocialAPI
@@ -17,7 +_,7 @@
     public static Terraria.Social.Base.WorkshopSocialModule Workshop;
     public static ServerJoinRequestsManager JoinRequests;
     public static Terraria.Social.Base.PlatformSocialModule Platform;
-    private static List<ISocialModule> _modules;
+    private static List<ISocialModule> _modules = new List<ISocialModule>();
     public static SocialMode Mode => _mode;
     public static void Initialize(SocialMode? mode = null)
     {
@@ -46,9 +_,6 @@
             case SocialMode.Steam:
                 LoadSteam();
                 break;
-            case SocialMode.WeGame:
-                LoadWeGame();
-                break;
         }
 
         foreach (ISocialModule module in _modules)
@@ -62,7 +_,7 @@
         _modules.Reverse();
         foreach (ISocialModule module in _modules)
         {
-            module.Shutdown();
+        	module.Shutdown();
         }
     }
 
@@ -106,25 +_,5 @@
         {
             Network = LoadModule<Terraria.Social.Steam.NetClientSocialModule>();
         }
-
-        WeGameHelper.WriteDebugString("LoadSteam modules");
-    }
-
-    private static void LoadWeGame()
-    {
-        LoadModule<Terraria.Social.WeGame.CoreSocialModule>();
-        Cloud = LoadModule<Terraria.Social.WeGame.CloudSocialModule>();
-        Friends = LoadModule<Terraria.Social.WeGame.FriendsSocialModule>();
-        Overlay = LoadModule<Terraria.Social.WeGame.OverlaySocialModule>();
-        if (Main.dedServ)
-        {
-            Network = LoadModule<Terraria.Social.WeGame.NetServerSocialModule>();
-        }
-        else
-        {
-            Network = LoadModule<Terraria.Social.WeGame.NetClientSocialModule>();
-        }
-
-        WeGameHelper.WriteDebugString("LoadWeGame modules");
     }
 }
