--- src/Terraria/ReLogic/ReLogic.csproj
+++ src/TerraAngel/ReLogic/ReLogic.csproj
@@ -7,20 +_,14 @@
     <Company>Re-Logic</Company>
     <Company>Copyright © Re-Logic 2017</Company>
     <RootNamespace>ReLogic</RootNamespace>
+    <TargetFramework>net7.0</TargetFramework>
+    <Platforms>AnyCPU</Platforms>
   </PropertyGroup>
   <ItemGroup>
-    <Reference Include="Ionic.Zip.CF" />
-    <Reference Include="Microsoft.Xna.Framework" />
-    <Reference Include="Microsoft.Xna.Framework.Game" />
-    <Reference Include="Microsoft.Xna.Framework.Graphics" />
-    <Reference Include="Newtonsoft.Json" />
-    <Reference Include="SteelSeriesEngineWrapper" />
-    <Reference Include="System" />
-    <Reference Include="System.Core" />
-    <Reference Include="System.Drawing" />
-    <Reference Include="System.Windows.Forms" />
+    <PackageReference Include="DotNetZip" Version="1.16.0" />
+    <PackageReference Include="Newtonsoft.Json" Version="13.0.1" />
   </ItemGroup>
   <ItemGroup>
-    <EmbeddedResource Include="Microsoft/**" />
+    <ProjectReference Include="..\..\..\TNA\FNA.Core.csproj" />
   </ItemGroup>
 </Project>
