--- src/Terraria/ReLogic/OS/Windows/Clipboard.cs
+++ src/TerraAngel/ReLogic/OS/Windows/Clipboard.cs
@@ -1,6 +_,5 @@
 using System;
 using System.Threading;
-using System.Windows.Forms;
 using ReLogic.OS.Base;
 
 namespace ReLogic.OS.Windows;
@@ -20,10 +_,7 @@
 
         try
         {
-            InvokeInStaThread(delegate
-            {
-                System.Windows.Forms.Clipboard.SetText(text);
-            });
+            SDL2.SDL.SDL_SetClipboardText(text);
         }
         catch
         {
@@ -35,7 +_,7 @@
     {
         try
         {
-            return InvokeInStaThread(() => System.Windows.Forms.Clipboard.GetText());
+            return SDL2.SDL.SDL_GetClipboardText();
         }
         catch
         {
@@ -43,24 +_,6 @@
             Console.WriteLine("Failed to get clipboard contents!");
             return result;
         }
-    }
-
-    private static T InvokeInStaThread<T>(Func<T> callback)
-    {
-        if (GetApartmentStateSafely() == ApartmentState.STA)
-        {
-            return callback();
-        }
-
-        T result = default(T);
-        Thread thread = new Thread((ThreadStart)delegate
-        {
-            result = callback();
-        });
-        thread.SetApartmentState(ApartmentState.STA);
-        thread.Start();
-        thread.Join();
-        return result;
     }
 
     private static void InvokeInStaThread(Action callback)
