--- src/Terraria/ReLogic/OS/Windows/NativeMethods.cs
+++ src/TerraAngel/ReLogic/OS/Windows/NativeMethods.cs
@@ -1,6 +_,5 @@
 using System;
 using System.Runtime.InteropServices;
-using System.Windows.Forms;
 
 namespace ReLogic.OS.Windows;
 internal static class NativeMethods
@@ -75,14 +_,6 @@
         DesktopVertRes = 117
     }
 
-    [DllImport("user32.dll", CharSet = CharSet.Unicode)]
-    public static extern IntPtr CallWindowProc(IntPtr lpPrevWndFunc, IntPtr hWnd, int msg, IntPtr wParam, IntPtr lParam);
-    [DllImport("user32.dll", CharSet = CharSet.Unicode)]
-    public static extern int SetWindowLong(IntPtr hWnd, int nIndex, int dwNewLong);
-    [DllImport("user32.dll", CharSet = CharSet.Unicode)]
-    public static extern IntPtr DefWindowProc(IntPtr hWnd, int msg, IntPtr wParam, IntPtr lParam);
-    [DllImport("user32.dll", CallingConvention = CallingConvention.StdCall, CharSet = CharSet.Unicode)]
-    public static extern bool TranslateMessage(ref Message message);
     [DllImport("user32.dll", CharSet = CharSet.Unicode)]
     public static extern IntPtr GetForegroundWindow();
     [DllImport("kernel32.dll", CharSet = CharSet.Unicode)]
