--- src/Terraria/Terraria/Social/Steam/CoreSocialModule.cs
+++ src/TerraAngel/Terraria/Social/Steam/CoreSocialModule.cs
@@ -1,6 +_,5 @@
 using System;
 using System.Threading;
-using System.Windows.Forms;
 using ReLogic.OS;
 using Steamworks;
 using Terraria.Localization;
@@ -31,7 +_,7 @@
 
         if (!SteamAPI.Init())
         {
-            MessageBox.Show(Language.GetTextValue("Error.LaunchFromSteam"), Language.GetTextValue("Error.Error"));
+            SDL2.SDL.SDL_ShowSimpleMessageBox(SDL2.SDL.SDL_MessageBoxFlags.SDL_MESSAGEBOX_ERROR, "Terraria: Error", Language.GetTextValue("Error.LaunchFromSteam"), Main.instance.Window.Handle);
             Environment.Exit(1);
         }
 
@@ -122,6 +_,6 @@
 
     public void OnOverlayActivated(GameOverlayActivated_t result)
     {
-        Main.instance.IsMouseVisible = result.m_bActive == 1;
+        Main.instance.IsMouseVisible = TerraAngel.Hooks.Hooks.MiscHooks.ShouldMouseBeVisible(result.m_bActive == 1);
     }
 }
