--- src/Terraria/Terraria/Localization/LanguageManager.cs
+++ src/TerraAngel/Terraria/Localization/LanguageManager.cs
@@ -92,7 +_,7 @@
     private string[] GetLanguageFilesForCulture(GameCulture culture)
     {
         Assembly.GetExecutingAssembly();
-        return Array.FindAll(typeof(Program).Assembly.GetManifestResourceNames(), (string element) => element.StartsWith("Terraria.Localization.Content." + culture.CultureInfo.Name) && element.EndsWith(".json"));
+        return Array.FindAll(typeof(Program).Assembly.GetManifestResourceNames(), (string element) => element.StartsWith("Terraria.Localization.Content." + culture.CultureInfo.Name.Replace('-', '_')) && element.EndsWith(".json"));
     }
 
     public void SetLanguage(GameCulture culture)
@@ -218,14 +_,21 @@
     public void LoadLanguageFromFileTextCsv(string fileText)
     {
         using TextReader reader = new StringReader(fileText);
-        using CsvReader csvReader = new CsvReader(reader);
-        csvReader.Configuration.HasHeaderRecord = true;
-        if (!csvReader.ReadHeader())
+        using CsvReader csvReader = new CsvReader(reader, new CsvHelper.Configuration.CsvConfiguration(System.Globalization.CultureInfo.InvariantCulture) { HasHeaderRecord = true, });
+
+        try
+        {
+            if (!csvReader.ReadHeader())
+            {
+                return;
+            }
+        }
+        catch (Exception)
         {
             return;
         }
 
-        string[] fieldHeaders = csvReader.FieldHeaders;
+        string[] fieldHeaders = csvReader.HeaderRecord;
         int num = -1;
         int num2 = -1;
         for (int i = 0; i < fieldHeaders.Length; i++)
@@ -250,7 +_,7 @@
         int num3 = Math.Max(num, num2) + 1;
         while (csvReader.Read())
         {
-            string[] currentRecord = csvReader.CurrentRecord;
+            string[] currentRecord = csvReader.GetRecord<string[]>();
             if (currentRecord.Length >= num3)
             {
                 string text2 = currentRecord[num];
