--- src/Terraria/Terraria/WorldBuilding/GenBase.cs
+++ src/TerraAngel/Terraria/WorldBuilding/GenBase.cs
@@ -6,7 +_,7 @@
     {
         public delegate bool CustomPerUnitAction(int x, int y, params object[] args);
         protected static UnifiedRandom _random => WorldGen.genRand;
-        protected static Tile[,] _tiles => Main.tile;
+        protected static Tilemap _tiles => Main.tile;
         protected static int _worldWidth => Main.maxTilesX;
         protected static int _worldHeight => Main.maxTilesY;
     }
