--- src/Terraria/Terraria/GameContent/UI/Chat/ItemTagHandler.cs
+++ src/TerraAngel/Terraria/GameContent/UI/Chat/ItemTagHandler.cs
@@ -7,9 +_,9 @@
 namespace Terraria.GameContent.UI.Chat;
 public class ItemTagHandler : ITagHandler
 {
-    private class ItemSnippet : TextSnippet
+    public class ItemSnippet : TextSnippet
     {
-        private Item _item;
+        public Item _item;
         public ItemSnippet(Item item)
         {
             _item = item;
