--- src/Terraria/Terraria/Graphics/Light/LightingEngine.cs
+++ src/TerraAngel/Terraria/Graphics/Light/LightingEngine.cs
@@ -37,6 +_,11 @@
     private Rectangle _workingProcessedArea;
     private readonly Stopwatch _timer = new Stopwatch();
     private EngineState _state;
+    public int BlurPassCount
+    {
+        get => _activeLightMap.BlurPassCount;
+        set => _activeLightMap.BlurPassCount = (_workingLightMap.BlurPassCount = value);
+    }
     public void AddLight(int x, int y, Vector3 color)
     {
         _perFrameLights.Add(new PerFrameLight(new Point(x, y), color));
@@ -198,7 +_,7 @@
         workingLightMap.LightDecayThroughAir *= Player.airLightDecay;
         workingLightMap.LightDecayThroughSolid *= Player.solidLightDecay;
     }
-
+    
     private void ApplyPerFrameLights()
     {
         for (int i = 0; i < _perFrameLights.Count; i++)
@@ -213,7 +_,11 @@
             }
         }
 
+        if (Lighting.AbleToProcessPerFrameLights)
+        {
+            Lighting.AbleToProcessPerFrameLights = false;
-        _perFrameLights.Clear();
+            _perFrameLights.Clear();
+        }
     }
 
     public void Rebuild()
