--- src/Terraria/Terraria/Graphics/Light/LightMap.cs
+++ src/TerraAngel/Terraria/Graphics/Light/LightMap.cs
@@ -1,3 +_,4 @@
+using System.Runtime.CompilerServices;
 using Microsoft.Xna.Framework;
 using ReLogic.Threading;
 using Terraria.Utilities;
@@ -204,7 +_,7 @@
                     break;
                 case LightMaskMode.Water:
                     {
-                        float num = (float)_random.WithModifier((ulong)i).Next(98, 100) / 100f;
+                        float num = 99f / 100f;
                         if (!flag)
                         {
                             zero.X *= LightDecayThroughWater.X * num;
@@ -244,6 +_,7 @@
         }
     }
 
+    [MethodImpl(MethodImplOptions.AggressiveInlining)]
     private int IndexOf(int x, int y)
     {
         return x * Height + y;
