--- src/Terraria/Terraria/Audio/ActiveSound.cs
+++ src/TerraAngel/Terraria/Audio/ActiveSound.cs
@@ -1,3 +_,4 @@
+using System.Threading;
 using Microsoft.Xna.Framework;
 using Microsoft.Xna.Framework.Audio;
 
@@ -14,11 +_,14 @@
     public bool IsPlaying => Sound.State == SoundState.Playing;
     public ActiveSound(SoundStyle style, Vector2 position)
     {
+        lock (SoundEngine.AudioLock)
+        {
-        Position = position;
+            Position = position;
-        Volume = 1f;
+            Volume = 1f;
-        IsGlobal = false;
+            IsGlobal = false;
-        Style = style;
+            Style = style;
-        Play();
+            Play();
+        }
     }
 
     public ActiveSound(SoundStyle style)
