--- src/Terraria/Terraria/Audio/ActiveSound.cs
+++ src/TerraAngel/Terraria/Audio/ActiveSound.cs
@@ -1,3 +_,4 @@
+using System.Threading;
 using Microsoft.Xna.Framework;
 using Microsoft.Xna.Framework.Audio;
 
@@ -17,32 +_,41 @@
     public bool IsPlaying => Sound.State == SoundState.Playing;
     public ActiveSound(SoundStyle style, Vector2 position)
     {
+        lock (SoundEngine.AudioLock)
+        {
         Position = position;
-        Volume = 1f;
         Pitch = style.PitchVariance;
+        Volume = 1f;
         IsGlobal = false;
         Style = style;
         Play();
+        }
     }
 
     public ActiveSound(SoundStyle style)
     {
+        lock (SoundEngine.AudioLock)
+        {
         Position = Vector2.Zero;
         Volume = 1f;
         Pitch = style.PitchVariance;
         IsGlobal = true;
         Style = style;
         Play();
+        }
     }
 
     public ActiveSound(SoundStyle style, Vector2 position, LoopedPlayCondition condition)
     {
+        lock (SoundEngine.AudioLock)
+        {
         Position = position;
         Volume = 1f;
         Pitch = style.PitchVariance;
         IsGlobal = false;
         Style = style;
         PlayLooped(condition);
+        }
     }
 
     private void Play()
@@ -95,7 +_,7 @@
 
     public void Update()
     {
-        if (Sound == null)
+        if (Sound == null || Sound != null && Sound.IsDisposed)
         {
             return;
         }
