--- src/Terraria/Terraria/Audio/LegacyAudioSystem.cs
+++ src/TerraAngel/Terraria/Audio/LegacyAudioSystem.cs
@@ -119,7 +_,7 @@
         Engine = new AudioEngine("Content\\TerrariaMusic.xgs");
         SoundBank = new SoundBank(Engine, "Content\\Sound Bank.xsb");
         Engine.Update();
-        WaveBank = new WaveBank(Engine, "Content\\Wave Bank.xwb", 0, 512);
+        WaveBank = new WaveBank(Engine, "Content\\Wave Bank.xwb");
         Engine.Update();
         AudioTracks = new IAudioTrack[91];
         TrackNamesByIndex = new Dictionary<int, string>();
