--- src/Terraria/Terraria/Audio/SoundPlayer.cs
+++ src/TerraAngel/Terraria/Audio/SoundPlayer.cs
@@ -1,4 +_,5 @@
 using System.Collections.Generic;
+using System.Threading;
 using Microsoft.Xna.Framework;
 using ReLogic.Utilities;
 
@@ -92,19 +_,22 @@
 
     public void Update()
     {
-        foreach (SlotVector<ActiveSound>.ItemPair item in (IEnumerable<SlotVector<ActiveSound>.ItemPair>)_trackedSounds)
+        lock (SoundEngine.AudioLock)
         {
-            try
+            foreach (SlotVector<ActiveSound>.ItemPair item in (IEnumerable<SlotVector<ActiveSound>.ItemPair>)_trackedSounds)
             {
+                try
+                {
-                item.Value.Update();
+                    item.Value.Update();
-                if (!item.Value.IsPlaying)
+                    if (!item.Value.IsPlaying)
+                    {
+                        _trackedSounds.Remove(item.Id);
+                    }
+                }
+                catch
                 {
                     _trackedSounds.Remove(item.Id);
                 }
-            }
-            catch
-            {
-                _trackedSounds.Remove(item.Id);
             }
         }
     }
