--- src/Terraria/Terraria/Audio/SoundEngine.cs
+++ src/TerraAngel/Terraria/Audio/SoundEngine.cs
@@ -1,5 +_,6 @@
 using System;
 using System.IO;
+using System.Threading;
 using Microsoft.Xna.Framework;
 using Microsoft.Xna.Framework.Audio;
 using ReLogic.Utilities;
@@ -7,6 +_,7 @@
 namespace Terraria.Audio;
 public static class SoundEngine
 {
+    public static readonly object AudioLock = new object();
     public static LegacySoundPlayer LegacySoundPlayer;
     public static SoundPlayer SoundPlayer;
     public static bool AreSoundsPaused;
@@ -93,8 +_,10 @@
         {
             return null;
         }
-
+        lock (AudioLock)
+        {
         return LegacySoundPlayer.PlaySound(type, x, y, Style, volumeScale, pitchOffset);
+        }
     }
 
     public static ActiveSound GetActiveSound(SlotId id)
