--- src/Terraria/Terraria/Graphics/Effects/FilterManager.cs
+++ src/TerraAngel/Terraria/Graphics/Effects/FilterManager.cs
@@ -155,7 +_,8 @@
             renderTarget2D = screenTarget2;
             graphicsDevice.SetRenderTarget(renderTarget2D);
             graphicsDevice.Clear(clearColor);
+            Matrix.Invert(Main.GameViewMatrix.EffectMatrix, out Matrix inverseEffectMatrix);
-            Main.spriteBatch.Begin(SpriteSortMode.Immediate, BlendState.AlphaBlend, SamplerState.LinearClamp, DepthStencilState.Default, RasterizerState.CullNone, null, Matrix.Invert(Main.GameViewMatrix.EffectMatrix));
+            Main.spriteBatch.Begin(SpriteSortMode.Immediate, BlendState.AlphaBlend, SamplerState.LinearClamp, DepthStencilState.Default, RasterizerState.CullNone, null, inverseEffectMatrix);
             Main.spriteBatch.Draw(renderTarget2D2, Vector2.Zero, Color.White);
             Main.spriteBatch.End();
             renderTarget2D2 = screenTarget2;
