--- src/Terraria/Terraria/UI/ChestUI.cs
+++ src/TerraAngel/Terraria/UI/ChestUI.cs
@@ -62,7 +_,7 @@
         if (Main.player[Main.myPlayer].chest != -1 && !Main.recBigList)
         {
             Main.inventoryScale = 0.755f;
-            if (Utils.FloatIntersect(Main.mouseX, Main.mouseY, 0f, 0f, 73f, Main.instance.invBottom, 560f * Main.inventoryScale, 224f * Main.inventoryScale))
+            if (Utils.FloatIntersect(Main.mouseX, Main.mouseY, 0f, 0f, 73f, Main.instance.invBottom + 9f, 560f * Main.inventoryScale, 224f * Main.inventoryScale + 9f))
             {
                 Main.player[Main.myPlayer].mouseInterface = true;
             }
@@ -108,7 +_,7 @@
                 text += "|";
             }
 
-            Main.instance.DrawWindowsIMEPanel(new Vector2(120f, 518f));
+            Main.instance.DrawWindowsIMEPanel(new Vector2(120f, 518f + 9f));
         }
         else if (player.chest > -1)
         {
@@ -167,7 +_,7 @@
         lineAmount++;
         for (int i = 0; i < lineAmount; i++)
         {
-            ChatManager.DrawColorCodedStringWithShadow(spritebatch, FontAssets.MouseText.Value, text, new Vector2(504f, Main.instance.invBottom + i * 26), color, 0f, Vector2.Zero, Vector2.One, -1f, 1.5f);
+            ChatManager.DrawColorCodedStringWithShadow(spritebatch, FontAssets.MouseText.Value, text, new Vector2(504f, Main.instance.invBottom + i * 26 + 9f), color, 0f, Vector2.Zero, Vector2.One, -1f, 1.5f);
         }
     }
 
@@ -175,7 +_,7 @@
     {
         for (int i = 0; i < ButtonID.Count; i++)
         {
-            DrawButton(spritebatch, i, 506, Main.instance.invBottom + 40);
+            DrawButton(spritebatch, i, 506, Main.instance.invBottom + 40 + 9);
         }
     }
 
@@ -366,7 +_,7 @@
         }
 
         Main.inventoryScale = 0.755f;
-        if (Utils.FloatIntersect(Main.mouseX, Main.mouseY, 0f, 0f, 73f, Main.instance.invBottom, 560f * Main.inventoryScale, 224f * Main.inventoryScale) && !PlayerInput.IgnoreMouseInterface)
+        if (Utils.FloatIntersect(Main.mouseX, Main.mouseY, 0f, 0f, 73f, Main.instance.invBottom + 9f, 560f * Main.inventoryScale, 224f * Main.inventoryScale + 9f) && !PlayerInput.IgnoreMouseInterface)
         {
             player.mouseInterface = true;
         }
@@ -376,7 +_,7 @@
             for (int j = 0; j < 4; j++)
             {
                 int num = (int)(73f + (float)(i * 56) * Main.inventoryScale);
-                int num2 = (int)((float)Main.instance.invBottom + (float)(j * 56) * Main.inventoryScale);
+                int num2 = (int)((float)Main.instance.invBottom + (float)(j * 56) * Main.inventoryScale) + 9;
                 int slot = i + j * 10;
                 new Color(100, 100, 100, 100);
                 if (Utils.FloatIntersect(Main.mouseX, Main.mouseY, 0f, 0f, num, num2, (float)TextureAssets.InventoryBack.Width() * Main.inventoryScale, (float)TextureAssets.InventoryBack.Height() * Main.inventoryScale) && !PlayerInput.IgnoreMouseInterface)
