--- src/Terraria/Terraria/UI/FancyErrorPrinter.cs
+++ src/TerraAngel/Terraria/UI/FancyErrorPrinter.cs
@@ -2,7 +_,6 @@
 using System.Collections.Generic;
 using System.IO;
 using System.Text;
-using System.Windows.Forms;
 using ReLogic.Content;
 using ReLogic.OS;
 
@@ -140,9 +_,9 @@
 
     private static void ShowTheBox(string preparedMessage)
     {
-        if (Platform.IsWindows && !Main.dedServ)
+        if (!Main.dedServ)
         {
-            MessageBox.Show(preparedMessage, "Error", MessageBoxButtons.OK, MessageBoxIcon.Hand);
+            SDL2.SDL.SDL_ShowSimpleMessageBox(SDL2.SDL.SDL_MessageBoxFlags.SDL_MESSAGEBOX_ERROR, "Terraria: Error", preparedMessage, Main.instance.Window.Handle);
         }
     }
 }
