--- src/Terraria/Terraria/UI/InGameNotificationsTracker.cs
+++ src/TerraAngel/Terraria/UI/InGameNotificationsTracker.cs
@@ -3,7 +_,6 @@
 using Microsoft.Xna.Framework.Graphics;
 using Terraria.Achievements;
 using Terraria.GameInput;
-using Terraria.Social;
 using Terraria.Social.Base;
 
 namespace Terraria.UI;
@@ -12,9 +_,9 @@
     private static List<IInGameNotification> _notifications = new List<IInGameNotification>();
     public static void Initialize()
     {
-        Main.Achievements.OnAchievementCompleted += AddCompleted;
+        // Main.Achievements.OnAchievementCompleted += AddCompleted;
-        SocialAPI.JoinRequests.OnRequestAdded += JoinRequests_OnRequestAdded;
+        // SocialAPI.JoinRequests.OnRequestAdded += JoinRequests_OnRequestAdded;
-        SocialAPI.JoinRequests.OnRequestRemoved += JoinRequests_OnRequestRemoved;
+        // SocialAPI.JoinRequests.OnRequestRemoved += JoinRequests_OnRequestRemoved;
     }
 
     private static void JoinRequests_OnRequestAdded(UserJoinToServerRequest request)
