--- src/Terraria/Terraria/UI/ItemSlot.cs
+++ src/TerraAngel/Terraria/UI/ItemSlot.cs
@@ -477,9 +_,9 @@
             }
         }
 
-        if (Main.keyState.IsKeyDown(Main.FavoriteKey) && (canFavoriteAt[context] || (Main.drawingPlayerChat && canShareAt[context])))
+        if (Main.keyState.IsKeyDown(Main.FavoriteKey) && (canFavoriteAt[context] || (TerraAngel.ClientLoader.ChatWindow?.IsChatting ?? false && canShareAt[context])))
         {
-            if (item.type > 0 && item.stack > 0 && Main.drawingPlayerChat)
+            if (item.type > 0 && item.stack > 0 && (TerraAngel.ClientLoader.ChatWindow?.IsChatting ?? false))
             {
                 Main.cursorOverride = 2;
             }
@@ -594,7 +_,7 @@
     public static void LeftClick(Item[] inv, int context = 0, int slot = 0)
     {
         Player player = Main.player[Main.myPlayer];
-        bool flag = Main.mouseLeftRelease && Main.mouseLeft;
+        bool flag = InputSystem.LeftMousePressed;
         if (flag)
         {
             if (OverrideLeftClick(inv, context, slot))
