--- src/Terraria/Terraria/UI/NetDiagnosticsUI.cs
+++ src/TerraAngel/Terraria/UI/NetDiagnosticsUI.cs
@@ -55,11 +_,13 @@
         public void CountReadMessage(int messageId, int messageLength)
         {
             _counterByMessageId[messageId].CountReadMessage(messageLength);
+            TerraAngel.Client.ClientWindows.StatsWindow.CountReadMessage(messageLength);
         }
 
         public void CountSentMessage(int messageId, int messageLength)
         {
             _counterByMessageId[messageId].CountSentMessage(messageLength);
+            TerraAngel.Client.ClientWindows.StatsWindow.CountSentMessage(messageLength);
         }
 
         public void CountReadModuleMessage(int moduleMessageId, int messageLength)
