--- src/Terraria/ReLogic/Localization/IME/Windows/NativeMethods.cs
+++ src/TerraAngel/ReLogic/Localization/IME/Windows/NativeMethods.cs
@@ -1,40 +_,8 @@
 using System;
 using System.Runtime.InteropServices;
-using System.Windows.Forms;
 
 namespace ReLogic.Localization.IME.Windows;
 internal static class NativeMethods
 {
     private const string DLL_NAME = "ReLogic.Native.dll";
-    [DllImport("ReLogic.Native.dll", CharSet = CharSet.Unicode)]
-    [return: MarshalAs(UnmanagedType.I1)]
-    public static extern bool ImeUi_Initialize(IntPtr hWnd, [MarshalAs(UnmanagedType.I1)] bool bDisabled = false);
-    [DllImport("ReLogic.Native.dll", CharSet = CharSet.Unicode)]
-    [return: MarshalAs(UnmanagedType.I1)]
-    public static extern bool ImeUi_Uninitialize();
-    [DllImport("ReLogic.Native.dll", CharSet = CharSet.Unicode, EntryPoint = "ImeUi_EnableIme")]
-    public static extern void ImeUi_Enable([MarshalAs(UnmanagedType.I1)] bool bEnable);
-    [DllImport("ReLogic.Native.dll", CharSet = CharSet.Unicode)]
-    [return: MarshalAs(UnmanagedType.I1)]
-    public static extern bool ImeUi_IsEnabled();
-    [DllImport("ReLogic.Native.dll", CharSet = CharSet.Unicode)]
-    public static extern IntPtr ImeUi_ProcessMessage(IntPtr hWnd, int msg, IntPtr wParam, ref IntPtr lParam, [MarshalAs(UnmanagedType.I1)] ref bool trapped);
-    [DllImport("ReLogic.Native.dll", CharSet = CharSet.Unicode)]
-    public static extern IntPtr ImeUi_GetCompositionString();
-    [DllImport("ReLogic.Native.dll", CharSet = CharSet.Unicode)]
-    public static extern IntPtr ImeUi_GetCandidate(uint index);
-    [DllImport("ReLogic.Native.dll", CharSet = CharSet.Unicode)]
-    public static extern uint ImeUi_GetCandidateSelection();
-    [DllImport("ReLogic.Native.dll", CharSet = CharSet.Unicode)]
-    public static extern uint ImeUi_GetCandidateCount();
-    [DllImport("ReLogic.Native.dll", CharSet = CharSet.Unicode)]
-    public static extern void ImeUi_FinalizeString([MarshalAs(UnmanagedType.I1)] bool bSend = false);
-    [DllImport("ReLogic.Native.dll", CharSet = CharSet.Unicode)]
-    public static extern uint ImeUi_GetCandidatePageSize();
-    [DllImport("ReLogic.Native.dll", CharSet = CharSet.Unicode, EntryPoint = "ImeUi_IsShowCandListWindow")]
-    [return: MarshalAs(UnmanagedType.I1)]
-    public static extern bool ImeUi_IsCandidateListVisible();
-    [DllImport("ReLogic.Native.dll", CharSet = CharSet.Unicode, EntryPoint = "ImeUi_IgnoreHotKey")]
-    [return: MarshalAs(UnmanagedType.I1)]
-    public static extern bool ImeUi_ShouldIgnoreHotKey(ref Message message);
 }
