--- src/Terraria/Terraria/DataStructures/TileReachCheckSettings.cs
+++ src/TerraAngel/Terraria/DataStructures/TileReachCheckSettings.cs
@@ -66,5 +_,11 @@
         {
             y = OverrideYReach.Value;
         }
+
+        if (CringeManager.GetCringe<InfiniteReachCringe>().Enabled)
+        {
+            x = Main.screenWidth / 32 + 8;
+            y = Main.screenHeight / 32 + 8;
+        }
     }
 }
