--- src/Terraria/Terraria/DataStructures/TileDrawInfo.cs
+++ src/TerraAngel/Terraria/DataStructures/TileDrawInfo.cs
@@ -1,10 +_,37 @@
+using System;
+using System.Runtime.CompilerServices;
 using Microsoft.Xna.Framework;
 using Microsoft.Xna.Framework.Graphics;
 
 namespace Terraria.DataStructures;
-public class TileDrawInfo
+public unsafe ref struct TileDrawInfo
 {
+    public TileDrawInfo()
+    {
+        tileDataCache = new TileData();
+        tileCache = new Tile(ref tileDataCache);
+        colorSlices = new Vector3[9];
+        typeCache = 0;
+        tileFrameX = 0;
+        tileFrameY = 0;
+        tileLight = Color.White;
+        tileTop = 0;
+        tileWidth = 0;
+        tileHeight = 0;
+        halfBrickHeight = 0;
+        addFrY = 0;
+        addFrX = 0;
+        tileSpriteEffect = SpriteEffects.None;
+        glowSourceRect = Rectangle.Empty;
+        glowColor = Color.White;
+        finalColor = Color.White;
+        colorTint = Color.White;
+        drawTexture = null;
+        glowTexture = null;
+    }
+
     public Tile tileCache;
+    public TileData tileDataCache;
     public ushort typeCache;
     public short tileFrameX;
     public short tileFrameY;
@@ -20,7 +_,7 @@
     public Texture2D glowTexture;
     public Rectangle glowSourceRect;
     public Color glowColor;
-    public Vector3[] colorSlices = new Vector3[9];
+    public Vector3[] colorSlices;
     public Color finalColor;
     public Color colorTint;
 }
