--- src/Terraria/Terraria/GameContent/Liquid/LiquidRenderer.cs
+++ src/TerraAngel/Terraria/GameContent/Liquid/LiquidRenderer.cs
@@ -78,7 +_,7 @@
     private readonly UnifiedRandom _random = new UnifiedRandom();
     private Color[] _waveMask = new Color[1];
     private float _frameState;
-    private static Tile[,] Tiles => Main.tile;
+    private static NativeTileMap Tiles => Main.tile;
     public event Action<Color[], Rectangle> WaveFilters;
     public static void LoadContent()
     {
